<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 10 Coalescent | Mathematical Demography</title>
  <meta name="description" content="This is a book from the Mathematical Demography lecture notes." />
  <meta name="generator" content="bookdown 0.19 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 10 Coalescent | Mathematical Demography" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is a book from the Mathematical Demography lecture notes." />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 10 Coalescent | Mathematical Demography" />
  
  <meta name="twitter:description" content="This is a book from the Mathematical Demography lecture notes." />
  

<meta name="author" content="Josh R. Goldstein" />


<meta name="date" content="2020-12-22" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="fisher-wright.html"/>
<link rel="next" href="student-presentations.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<script src="libs/htmlwidgets-1.5.3/htmlwidgets.js"></script>
<script src="libs/viz-1.8.2/viz.js"></script>
<link href="libs/DiagrammeR-styles-0.2/styles.css" rel="stylesheet" />
<script src="libs/grViz-binding-1.0.6.1/grViz.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>


<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Random Demography</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction to Demographic Heterogeneity</a><ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#outline"><i class="fa fa-check"></i><b>2.1</b> Outline</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#part-i.-conceptual-introduction"><i class="fa fa-check"></i><b>2.2</b> Part I. Conceptual Introduction</a><ul>
<li class="chapter" data-level="2.2.1" data-path="intro.html"><a href="intro.html#what-is-demographic-heterogeneity"><i class="fa fa-check"></i><b>2.2.1</b> What is Demographic Heterogeneity?</a></li>
<li class="chapter" data-level="2.2.2" data-path="intro.html"><a href="intro.html#analogies"><i class="fa fa-check"></i><b>2.2.2</b> Analogies</a></li>
<li class="chapter" data-level="2.2.3" data-path="intro.html"><a href="intro.html#whats-new"><i class="fa fa-check"></i><b>2.2.3</b> What’s new?</a></li>
<li class="chapter" data-level="2.2.4" data-path="intro.html"><a href="intro.html#application"><i class="fa fa-check"></i><b>2.2.4</b> Application</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#part-ii.-formal-analysis"><i class="fa fa-check"></i><b>2.3</b> Part II. Formal Analysis</a><ul>
<li class="chapter" data-level="2.3.1" data-path="intro.html"><a href="intro.html#outline-1"><i class="fa fa-check"></i><b>2.3.1</b> Outline</a></li>
<li class="chapter" data-level="2.3.2" data-path="intro.html"><a href="intro.html#keyfitz-result"><i class="fa fa-check"></i><b>2.3.2</b> Keyfitz result</a></li>
<li class="chapter" data-level="2.3.3" data-path="intro.html"><a href="intro.html#us-mexico-example"><i class="fa fa-check"></i><b>2.3.3</b> US-Mexico Example</a></li>
<li class="chapter" data-level="2.3.4" data-path="intro.html"><a href="intro.html#the-origin-of-professor-wachters-poisson-exponential-model"><i class="fa fa-check"></i><b>2.3.4</b> The Origin of Professor Wachter’s Poisson-Exponential Model</a></li>
<li class="chapter" data-level="2.3.5" data-path="intro.html"><a href="intro.html#some-commentary"><i class="fa fa-check"></i><b>2.3.5</b> Some commentary</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="intro.html"><a href="intro.html#conclusions"><i class="fa fa-check"></i><b>2.4</b> Conclusions</a></li>
<li class="chapter" data-level="2.5" data-path="intro.html"><a href="intro.html#questions"><i class="fa fa-check"></i><b>2.5</b> Questions</a></li>
<li class="chapter" data-level="2.6" data-path="intro.html"><a href="intro.html#solutions"><i class="fa fa-check"></i><b>2.6</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="frailty.html"><a href="frailty.html"><i class="fa fa-check"></i><b>3</b> Multiplicative Fixed Frailty</a><ul>
<li class="chapter" data-level="3.1" data-path="frailty.html"><a href="frailty.html#outline-2"><i class="fa fa-check"></i><b>3.1</b> Outline</a><ul>
<li class="chapter" data-level="3.1.1" data-path="frailty.html"><a href="frailty.html#review-of-mortality-mathematics"><i class="fa fa-check"></i><b>3.1.1</b> Review of mortality mathematics</a></li>
<li class="chapter" data-level="3.1.2" data-path="frailty.html"><a href="frailty.html#application-what-is-ellx"><i class="fa fa-check"></i><b>3.1.2</b> Application: what is <span class="math inline">\(\ell&#39;(x)\)</span>?</a></li>
<li class="chapter" data-level="3.1.3" data-path="frailty.html"><a href="frailty.html#extending-keyfitz-to-mortality"><i class="fa fa-check"></i><b>3.1.3</b> Extending Keyfitz to mortality</a></li>
<li class="chapter" data-level="3.1.4" data-path="frailty.html"><a href="frailty.html#multiplicative-fixed-frailty"><i class="fa fa-check"></i><b>3.1.4</b> Multiplicative Fixed frailty</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="frailty.html"><a href="frailty.html#part-i.-results-from-fixed-frailty"><i class="fa fa-check"></i><b>3.2</b> Part I. Results from Fixed Frailty</a><ul>
<li class="chapter" data-level="3.2.1" data-path="frailty.html"><a href="frailty.html#a-simulation"><i class="fa fa-check"></i><b>3.2.1</b> A simulation</a></li>
<li class="chapter" data-level="3.2.2" data-path="frailty.html"><a href="frailty.html#lets-derive-pop-survival-note-bars-barell"><i class="fa fa-check"></i><b>3.2.2</b> Let’s derive pop survival (Note: <span class="math inline">\(\bar{s} = \bar{\ell}\)</span>)</a></li>
<li class="chapter" data-level="3.2.3" data-path="frailty.html"><a href="frailty.html#now-population-hazards-stepping-stones"><i class="fa fa-check"></i><b>3.2.3</b> Now population hazards (stepping stones)</a></li>
<li class="chapter" data-level="3.2.4" data-path="frailty.html"><a href="frailty.html#rodriguez-question"><i class="fa fa-check"></i><b>3.2.4</b> Rodriguez question</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="frailty.html"><a href="frailty.html#part-ii.-introduction-to-gamma-frailty"><i class="fa fa-check"></i><b>3.3</b> Part II. Introduction to Gamma Frailty</a><ul>
<li class="chapter" data-level="3.3.1" data-path="frailty.html"><a href="frailty.html#the-gamma-distribution"><i class="fa fa-check"></i><b>3.3.1</b> The Gamma distribution</a></li>
<li class="chapter" data-level="3.3.2" data-path="frailty.html"><a href="frailty.html#gamma-in-r"><i class="fa fa-check"></i><b>3.3.2</b> Gamma in R</a></li>
<li class="chapter" data-level="3.3.3" data-path="frailty.html"><a href="frailty.html#population-survival-of-gamma-frailty"><i class="fa fa-check"></i><b>3.3.3</b> Population Survival of Gamma Frailty</a></li>
<li class="chapter" data-level="3.3.4" data-path="frailty.html"><a href="frailty.html#interpreting-gamma-frailty-survival"><i class="fa fa-check"></i><b>3.3.4</b> Interpreting Gamma-frailty survival</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="frailty.html"><a href="frailty.html#conclusions-1"><i class="fa fa-check"></i><b>3.4</b> Conclusions</a></li>
<li class="chapter" data-level="3.5" data-path="frailty.html"><a href="frailty.html#questions-1"><i class="fa fa-check"></i><b>3.5</b> Questions</a></li>
<li class="chapter" data-level="3.6" data-path="frailty.html"><a href="frailty.html#solutions-1"><i class="fa fa-check"></i><b>3.6</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="gamma-frailty-with-applications.html"><a href="gamma-frailty-with-applications.html"><i class="fa fa-check"></i><b>4</b> Gamma Frailty with Applications</a><ul>
<li class="chapter" data-level="4.1" data-path="gamma-frailty-with-applications.html"><a href="gamma-frailty-with-applications.html#from-survival-to-hazards"><i class="fa fa-check"></i><b>4.1</b> From survival to hazards</a><ul>
<li class="chapter" data-level="4.1.1" data-path="gamma-frailty-with-applications.html"><a href="gamma-frailty-with-applications.html#what-happens-to-frailty-of-survivors"><i class="fa fa-check"></i><b>4.1.1</b> What happens to frailty of survivors?</a></li>
<li class="chapter" data-level="4.1.2" data-path="gamma-frailty-with-applications.html"><a href="gamma-frailty-with-applications.html#average-frailty-in-terms-of-survival"><i class="fa fa-check"></i><b>4.1.2</b> Average frailty in terms of survival</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="gamma-frailty-with-applications.html"><a href="gamma-frailty-with-applications.html#censoc-selection-and-observed-frailty"><i class="fa fa-check"></i><b>4.2</b> CenSoc: Selection and observed frailty</a><ul>
<li class="chapter" data-level="4.2.1" data-path="gamma-frailty-with-applications.html"><a href="gamma-frailty-with-applications.html#data"><i class="fa fa-check"></i><b>4.2.1</b> Data</a></li>
<li class="chapter" data-level="4.2.2" data-path="gamma-frailty-with-applications.html"><a href="gamma-frailty-with-applications.html#estimation"><i class="fa fa-check"></i><b>4.2.2</b> Estimation</a></li>
<li class="chapter" data-level="4.2.3" data-path="gamma-frailty-with-applications.html"><a href="gamma-frailty-with-applications.html#discuss-our-conclusions-and-possible-future-directions-to-follow."><i class="fa fa-check"></i><b>4.2.3</b> Discuss our conclusions and possible future directions to follow.</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="gamma-frailty-with-applications.html"><a href="gamma-frailty-with-applications.html#questions-2"><i class="fa fa-check"></i><b>4.3</b> Questions</a></li>
<li class="chapter" data-level="4.4" data-path="gamma-frailty-with-applications.html"><a href="gamma-frailty-with-applications.html#solutions-2"><i class="fa fa-check"></i><b>4.4</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="convergence-and-cross-overs.html"><a href="convergence-and-cross-overs.html"><i class="fa fa-check"></i><b>5</b> Convergence and cross-overs</a><ul>
<li class="chapter" data-level="5.1" data-path="convergence-and-cross-overs.html"><a href="convergence-and-cross-overs.html#outline-3"><i class="fa fa-check"></i><b>5.1</b> Outline</a></li>
<li class="chapter" data-level="5.2" data-path="convergence-and-cross-overs.html"><a href="convergence-and-cross-overs.html#what-happens-to-mortality-disparities-at-older-ages"><i class="fa fa-check"></i><b>5.2</b> What happens to mortality disparities at older ages?</a><ul>
<li class="chapter" data-level="5.2.1" data-path="convergence-and-cross-overs.html"><a href="convergence-and-cross-overs.html#a-possible-null-model"><i class="fa fa-check"></i><b>5.2.1</b> A possible null-model</a></li>
<li class="chapter" data-level="5.2.2" data-path="convergence-and-cross-overs.html"><a href="convergence-and-cross-overs.html#a-result-5e---vaupel2014unobserved-eq-38"><i class="fa fa-check"></i><b>5.2.2</b> A result: (5E - <span class="citation">Vaupel and Missov (<span>2014</span>)</span> (Eq 38))</a></li>
<li class="chapter" data-level="5.2.3" data-path="convergence-and-cross-overs.html"><a href="convergence-and-cross-overs.html#inversion"><i class="fa fa-check"></i><b>5.2.3</b> Inversion</a></li>
<li class="chapter" data-level="5.2.4" data-path="convergence-and-cross-overs.html"><a href="convergence-and-cross-overs.html#an-example-1"><i class="fa fa-check"></i><b>5.2.4</b> An example</a></li>
<li class="chapter" data-level="5.2.5" data-path="convergence-and-cross-overs.html"><a href="convergence-and-cross-overs.html#application-1"><i class="fa fa-check"></i><b>5.2.5</b> Application</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="convergence-and-cross-overs.html"><a href="convergence-and-cross-overs.html#student-presentation"><i class="fa fa-check"></i><b>5.3</b> Student Presentation</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="mortality-plateaus.html"><a href="mortality-plateaus.html"><i class="fa fa-check"></i><b>6</b> Mortality plateaus</a><ul>
<li class="chapter" data-level="6.1" data-path="mortality-plateaus.html"><a href="mortality-plateaus.html#outline-4"><i class="fa fa-check"></i><b>6.1</b> Outline</a></li>
<li class="chapter" data-level="6.2" data-path="mortality-plateaus.html"><a href="mortality-plateaus.html#heterogeneity-slows-mortality-improvement"><i class="fa fa-check"></i><b>6.2</b> Heterogeneity slows mortality improvement</a><ul>
<li class="chapter" data-level="6.2.1" data-path="mortality-plateaus.html"><a href="mortality-plateaus.html#an-example-2"><i class="fa fa-check"></i><b>6.2.1</b> An example</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="mortality-plateaus.html"><a href="mortality-plateaus.html#conclusions-2"><i class="fa fa-check"></i><b>6.3</b> Conclusions</a></li>
<li class="chapter" data-level="6.4" data-path="mortality-plateaus.html"><a href="mortality-plateaus.html#hazards-and-plateaus-by-professor-ken-wachter"><i class="fa fa-check"></i><b>6.4</b> Hazards and Plateaus by Professor Ken Wachter</a><ul>
<li class="chapter" data-level="6.4.1" data-path="mortality-plateaus.html"><a href="mortality-plateaus.html#outline-5"><i class="fa fa-check"></i><b>6.4.1</b> Outline</a></li>
<li class="chapter" data-level="6.4.2" data-path="mortality-plateaus.html"><a href="mortality-plateaus.html#extremes-of-longevity-in-humans-and-other-species"><i class="fa fa-check"></i><b>6.4.2</b> Extremes of Longevity in Humans and Other Species</a></li>
<li class="chapter" data-level="6.4.3" data-path="mortality-plateaus.html"><a href="mortality-plateaus.html#gamma-gompertz-fixed-frailty-hazards"><i class="fa fa-check"></i><b>6.4.3</b> Gamma-Gompertz Fixed Frailty Hazards</a></li>
<li class="chapter" data-level="6.4.4" data-path="mortality-plateaus.html"><a href="mortality-plateaus.html#less-restrictive-frailty-models"><i class="fa fa-check"></i><b>6.4.4</b> Less Restrictive Frailty Models</a></li>
<li class="chapter" data-level="6.4.5" data-path="mortality-plateaus.html"><a href="mortality-plateaus.html#mutation-accumulation-gompertz-hazards-with-plateaus"><i class="fa fa-check"></i><b>6.4.5</b> Mutation Accumulation, Gompertz Hazards with Plateaus</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="mortality-plateaus.html"><a href="mortality-plateaus.html#key-points"><i class="fa fa-check"></i><b>6.5</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html"><a href="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html"><i class="fa fa-check"></i><b>7</b> Fertility Heterogeneity: Tempo Distortions and Distorted Tempo</a><ul>
<li class="chapter" data-level="7.1" data-path="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html"><a href="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html#outline-6"><i class="fa fa-check"></i><b>7.1</b> Outline</a></li>
<li class="chapter" data-level="7.2" data-path="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html"><a href="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html#introduction-1"><i class="fa fa-check"></i><b>7.2</b> Introduction</a><ul>
<li class="chapter" data-level="7.2.1" data-path="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html"><a href="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html#fertility-postponement-a-very-simple-example"><i class="fa fa-check"></i><b>7.2.1</b> Fertility postponement, a very simple example</a></li>
<li class="chapter" data-level="7.2.2" data-path="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html"><a href="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html#continuous-postponement-a-shiny-simulation"><i class="fa fa-check"></i><b>7.2.2</b> Continuous postponement, a shiny simulation</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html"><a href="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html#period-shifts-bongaarts-and-feeneys-model"><i class="fa fa-check"></i><b>7.3</b> Period Shifts: Bongaarts and Feeney’s model</a><ul>
<li class="chapter" data-level="7.3.1" data-path="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html"><a href="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html#a-derivation-due-to-rodriguez-note-not-sure-what-the-reference-is-here"><i class="fa fa-check"></i><b>7.3.1</b> A derivation: due to Rodriguez <em>NOTE: not sure what the reference is here</em></a></li>
<li class="chapter" data-level="7.3.2" data-path="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html"><a href="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html#tempo-adjusted-tfr"><i class="fa fa-check"></i><b>7.3.2</b> Tempo-adjusted TFR:</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html"><a href="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html#an-application-to-the-united-states"><i class="fa fa-check"></i><b>7.4</b> An Application to the United States</a><ul>
<li class="chapter" data-level="7.4.1" data-path="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html"><a href="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html#conclusions-3"><i class="fa fa-check"></i><b>7.4.1</b> Conclusions</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html"><a href="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html#two-americas"><i class="fa fa-check"></i><b>7.5</b> Two Americas</a><ul>
<li class="chapter" data-level="7.5.1" data-path="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html"><a href="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html#mixture"><i class="fa fa-check"></i><b>7.5.1</b> Mixture</a></li>
<li class="chapter" data-level="7.5.2" data-path="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html"><a href="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html#an-algorithm-for-tempo-adjustment-of-mixtures"><i class="fa fa-check"></i><b>7.5.2</b> An algorithm for tempo adjustment of mixtures</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html"><a href="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html#conclusions-4"><i class="fa fa-check"></i><b>7.6</b> Conclusions</a><ul>
<li class="chapter" data-level="7.6.1" data-path="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html"><a href="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html#caveats"><i class="fa fa-check"></i><b>7.6.1</b> Caveats</a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html"><a href="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html#questions-3"><i class="fa fa-check"></i><b>7.7</b> Questions</a></li>
<li class="chapter" data-level="7.8" data-path="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html"><a href="fertility-heterogeneity-tempo-distortions-and-distorted-tempo.html#solutions-3"><i class="fa fa-check"></i><b>7.8</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="branching-processes.html"><a href="branching-processes.html"><i class="fa fa-check"></i><b>8</b> Branching Processes</a><ul>
<li class="chapter" data-level="8.1" data-path="branching-processes.html"><a href="branching-processes.html#outline-7"><i class="fa fa-check"></i><b>8.1</b> Outline</a></li>
<li class="chapter" data-level="8.2" data-path="branching-processes.html"><a href="branching-processes.html#motivation"><i class="fa fa-check"></i><b>8.2</b> Motivation</a><ul>
<li class="chapter" data-level="8.2.1" data-path="branching-processes.html"><a href="branching-processes.html#very-brief-history-of-branching-processes"><i class="fa fa-check"></i><b>8.2.1</b> Very brief history of Branching Processes</a></li>
<li class="chapter" data-level="8.2.2" data-path="branching-processes.html"><a href="branching-processes.html#applicability-to-the-coronavirus-yes-and-no."><i class="fa fa-check"></i><b>8.2.2</b> Applicability to the Coronavirus? Yes and no.</a></li>
<li class="chapter" data-level="8.2.3" data-path="branching-processes.html"><a href="branching-processes.html#simulated-examples-and-the-questions-they-raise"><i class="fa fa-check"></i><b>8.2.3</b> Simulated examples and the questions they raise</a></li>
<li class="chapter" data-level="8.2.4" data-path="branching-processes.html"><a href="branching-processes.html#what-is-a-bienayme-galton-watson-branching-process"><i class="fa fa-check"></i><b>8.2.4</b> What is a (Bienayme)-Galton-Watson branching process?</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="branching-processes.html"><a href="branching-processes.html#simulations-of-parents-having-children"><i class="fa fa-check"></i><b>8.3</b> Simulations of parents having children</a></li>
<li class="chapter" data-level="8.4" data-path="branching-processes.html"><a href="branching-processes.html#the-probability-generating-function-our-mathematical-tool"><i class="fa fa-check"></i><b>8.4</b> The Probability Generating Function: Our mathematical tool</a></li>
<li class="chapter" data-level="8.5" data-path="branching-processes.html"><a href="branching-processes.html#extinction"><i class="fa fa-check"></i><b>8.5</b> Extinction</a></li>
<li class="chapter" data-level="8.6" data-path="branching-processes.html"><a href="branching-processes.html#good-and-bad-set-ups-for-branching-process"><i class="fa fa-check"></i><b>8.6</b> Good and bad set-ups for branching process</a></li>
<li class="chapter" data-level="8.7" data-path="branching-processes.html"><a href="branching-processes.html#the-distribution-of-offspring-of-all-generations"><i class="fa fa-check"></i><b>8.7</b> The distribution of offspring of all generations</a><ul>
<li class="chapter" data-level="8.7.1" data-path="branching-processes.html"><a href="branching-processes.html#proving-mathbbe-z_n-mn"><i class="fa fa-check"></i><b>8.7.1</b> Proving <span class="math inline">\(\mathbb{E} Z_n = m^n\)</span></a></li>
<li class="chapter" data-level="8.7.2" data-path="branching-processes.html"><a href="branching-processes.html#variance-result."><i class="fa fa-check"></i><b>8.7.2</b> Variance result.</a></li>
</ul></li>
<li class="chapter" data-level="8.8" data-path="branching-processes.html"><a href="branching-processes.html#geometric-offspring-distribution"><i class="fa fa-check"></i><b>8.8</b> Geometric offspring distribution</a><ul>
<li class="chapter" data-level="8.8.1" data-path="branching-processes.html"><a href="branching-processes.html#the-geometric-distributions-simple-pgf"><i class="fa fa-check"></i><b>8.8.1</b> The Geometric Distribution’s simple PGF</a></li>
<li class="chapter" data-level="8.8.2" data-path="branching-processes.html"><a href="branching-processes.html#a-plot-of-keyfitzs-numbers-for-generations-1-2-and-3.-is-it-exponential-for-k-0"><i class="fa fa-check"></i><b>8.8.2</b> A plot of Keyfitz’s numbers for generations 1, 2, and 3. Is it exponential for <span class="math inline">\(k &gt; 0\)</span>?</a></li>
</ul></li>
<li class="chapter" data-level="8.9" data-path="branching-processes.html"><a href="branching-processes.html#branching-processes-and-covid-19"><i class="fa fa-check"></i><b>8.9</b> Branching Processes and Covid-19</a></li>
<li class="chapter" data-level="8.10" data-path="branching-processes.html"><a href="branching-processes.html#problems"><i class="fa fa-check"></i><b>8.10</b> Problems</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="fisher-wright.html"><a href="fisher-wright.html"><i class="fa fa-check"></i><b>9</b> Fisher-Wright</a><ul>
<li class="chapter" data-level="9.1" data-path="fisher-wright.html"><a href="fisher-wright.html#outline-8"><i class="fa fa-check"></i><b>9.1</b> Outline</a></li>
<li class="chapter" data-level="9.2" data-path="fisher-wright.html"><a href="fisher-wright.html#parallel"><i class="fa fa-check"></i><b>9.2</b> Parallel</a><ul>
<li class="chapter" data-level="9.2.1" data-path="fisher-wright.html"><a href="fisher-wright.html#another-cell-phone-example"><i class="fa fa-check"></i><b>9.2.1</b> Another cell phone example</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="fisher-wright.html"><a href="fisher-wright.html#mutation"><i class="fa fa-check"></i><b>9.3</b> Mutation</a></li>
<li class="chapter" data-level="9.4" data-path="fisher-wright.html"><a href="fisher-wright.html#fixation"><i class="fa fa-check"></i><b>9.4</b> Fixation</a><ul>
<li class="chapter" data-level="9.4.1" data-path="fisher-wright.html"><a href="fisher-wright.html#probability-that-a-particular-line-will-fix"><i class="fa fa-check"></i><b>9.4.1</b> Probability that a particular line will “fix”</a></li>
<li class="chapter" data-level="9.4.2" data-path="fisher-wright.html"><a href="fisher-wright.html#expected-time-until-fixation"><i class="fa fa-check"></i><b>9.4.2</b> Expected time until fixation?</a></li>
<li class="chapter" data-level="9.4.3" data-path="fisher-wright.html"><a href="fisher-wright.html#path-to-fixation-a-measure-of-homogeneityheterogeneity"><i class="fa fa-check"></i><b>9.4.3</b> Path to fixation: a measure of homogeneity/heterogeneity</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="fisher-wright.html"><a href="fisher-wright.html#baby-names"><i class="fa fa-check"></i><b>9.5</b> Baby Names</a><ul>
<li class="chapter" data-level="9.5.1" data-path="fisher-wright.html"><a href="fisher-wright.html#basic-idea"><i class="fa fa-check"></i><b>9.5.1</b> Basic idea</a></li>
<li class="chapter" data-level="9.5.2" data-path="fisher-wright.html"><a href="fisher-wright.html#fisher-wright-simulation-of-baby-names-hahn-and-bentley"><i class="fa fa-check"></i><b>9.5.2</b> Fisher-Wright simulation of Baby Names (Hahn and Bentley)</a></li>
<li class="chapter" data-level="9.5.3" data-path="fisher-wright.html"><a href="fisher-wright.html#drawing-their-picture-with-simulation"><i class="fa fa-check"></i><b>9.5.3</b> Drawing their picture with simulation</a></li>
<li class="chapter" data-level="9.5.4" data-path="fisher-wright.html"><a href="fisher-wright.html#fw-babyname-simulation-of-equilibrium-frequencies"><i class="fa fa-check"></i><b>9.5.4</b> FW babyname simulation of equilibrium frequencies</a></li>
</ul></li>
<li class="chapter" data-level="9.6" data-path="fisher-wright.html"><a href="fisher-wright.html#now-we-can-simulate-babyynmes"><i class="fa fa-check"></i><b>9.6</b> Now we can simulate babyynmes</a></li>
<li class="chapter" data-level="9.7" data-path="fisher-wright.html"><a href="fisher-wright.html#conclusions-5"><i class="fa fa-check"></i><b>9.7</b> Conclusions</a><ul>
<li class="chapter" data-level="9.7.1" data-path="fisher-wright.html"><a href="fisher-wright.html#some-potential-criticism"><i class="fa fa-check"></i><b>9.7.1</b> Some potential criticism</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="coalescent.html"><a href="coalescent.html"><i class="fa fa-check"></i><b>10</b> Coalescent</a><ul>
<li class="chapter" data-level="10.1" data-path="coalescent.html"><a href="coalescent.html#outline-9"><i class="fa fa-check"></i><b>10.1</b> Outline</a></li>
<li class="chapter" data-level="10.2" data-path="coalescent.html"><a href="coalescent.html#the-coalescent-expectations-of-the-past"><i class="fa fa-check"></i><b>10.2</b> The Coalescent: Expectations of the Past</a></li>
<li class="chapter" data-level="10.3" data-path="coalescent.html"><a href="coalescent.html#our-first-question-when-was-mrca"><i class="fa fa-check"></i><b>10.3</b> Our first question: When was MRCA?</a></li>
<li class="chapter" data-level="10.4" data-path="coalescent.html"><a href="coalescent.html#mutation-and-inference-of-tmrca-and-n"><i class="fa fa-check"></i><b>10.4</b> Mutation and inference of TMRCA and <span class="math inline">\(N\)</span></a><ul>
<li class="chapter" data-level="10.4.1" data-path="coalescent.html"><a href="coalescent.html#inference-of-population-size-simulation"><i class="fa fa-check"></i><b>10.4.1</b> Inference of population size (Simulation)</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="coalescent.html"><a href="coalescent.html#coalescence-of-a-sample-of-n-individuals"><i class="fa fa-check"></i><b>10.5</b> Coalescence of a sample of <span class="math inline">\(n\)</span> individuals</a></li>
<li class="chapter" data-level="10.6" data-path="coalescent.html"><a href="coalescent.html#an-application-of-coalescent-theory"><i class="fa fa-check"></i><b>10.6</b> An application of coalescent theory</a><ul>
<li class="chapter" data-level="10.6.1" data-path="coalescent.html"><a href="coalescent.html#coalesence-when-population-is-changing"><i class="fa fa-check"></i><b>10.6.1</b> Coalesence when population is changing</a></li>
</ul></li>
<li class="chapter" data-level="10.7" data-path="coalescent.html"><a href="coalescent.html#reconstruction-ancient-european-population-sizes-using-batinis-sample-of-mitochondrial-dna"><i class="fa fa-check"></i><b>10.7</b> Reconstruction Ancient European Population Sizes using Batini’s sample of Mitochondrial DNA"</a><ul>
<li class="chapter" data-level="10.7.1" data-path="coalescent.html"><a href="coalescent.html#summary-of-batini-et-al."><i class="fa fa-check"></i><b>10.7.1</b> Summary of Batini et al.</a></li>
<li class="chapter" data-level="10.7.2" data-path="coalescent.html"><a href="coalescent.html#ancient-population-estimates"><i class="fa fa-check"></i><b>10.7.2</b> Ancient population estimates</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="student-presentations.html"><a href="student-presentations.html"><i class="fa fa-check"></i><b>11</b> Student Presentations</a><ul>
<li class="chapter" data-level="11.1" data-path="student-presentations.html"><a href="student-presentations.html#mortality-crossovers"><i class="fa fa-check"></i><b>11.1</b> Mortality Crossovers</a><ul>
<li class="chapter" data-level="11.1.1" data-path="student-presentations.html"><a href="student-presentations.html#outline-10"><i class="fa fa-check"></i><b>11.1.1</b> Outline</a></li>
<li class="chapter" data-level="11.1.2" data-path="student-presentations.html"><a href="student-presentations.html#mortality-crossovers-reality-or-bad-data-coale1986mortality"><i class="fa fa-check"></i><b>11.1.2</b> <em>Mortality Crossovers: Reality or Bad Data?</em> <span class="citation">(Coale and Kisker <span>1986</span>)</span></a></li>
<li class="chapter" data-level="11.1.3" data-path="student-presentations.html"><a href="student-presentations.html#methods-for-evaluating-the-heterogeneity-of-aging-processes-in-human-populations-using-vital-statistics-data-explaining-the-blackwhite-mortality-crossover-by-a-model-of-mortality-selection-manton1981methods"><i class="fa fa-check"></i><b>11.1.3</b> <em>Methods for Evaluating the Heterogeneity of Aging Processes in Human Populations Using Vital Statistics Data: Explaining the Black/White Mortality Crossover by a Model of Mortality Selection</em> <span class="citation">(Manton and Stallard <span>1981</span>)</span></a></li>
<li class="chapter" data-level="11.1.4" data-path="student-presentations.html"><a href="student-presentations.html#censoc-mortality-crossovers"><i class="fa fa-check"></i><b>11.1.4</b> CenSoc Mortality Crossovers</a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="student-presentations.html"><a href="student-presentations.html#plateaus"><i class="fa fa-check"></i><b>11.2</b> Plateaus</a></li>
<li class="chapter" data-level="11.3" data-path="student-presentations.html"><a href="student-presentations.html#rising-inequality"><i class="fa fa-check"></i><b>11.3</b> Rising Inequality</a><ul>
<li class="chapter" data-level="11.3.1" data-path="student-presentations.html"><a href="student-presentations.html#trends-in-mortality-differentials-and-life-expectancy-for-male-social-security-covered-workers-by-socioeconomic-status-waldron2007trends"><i class="fa fa-check"></i><b>11.3.1</b> <em>Trends in mortality differentials and life expectancy for male social security-covered workers, by socioeconomic status</em> <span class="citation">(Waldron <span>2007</span>)</span></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="12" data-path="demographic-theory-and-covid-19.html"><a href="demographic-theory-and-covid-19.html"><i class="fa fa-check"></i><b>12</b> Demographic theory and COVID-19</a><ul>
<li class="chapter" data-level="12.1" data-path="demographic-theory-and-covid-19.html"><a href="demographic-theory-and-covid-19.html#outline-11"><i class="fa fa-check"></i><b>12.1</b> Outline</a></li>
<li class="chapter" data-level="12.2" data-path="demographic-theory-and-covid-19.html"><a href="demographic-theory-and-covid-19.html#motivating-questions"><i class="fa fa-check"></i><b>12.2</b> Motivating questions</a></li>
<li class="chapter" data-level="12.3" data-path="demographic-theory-and-covid-19.html"><a href="demographic-theory-and-covid-19.html#population-aging"><i class="fa fa-check"></i><b>12.3</b> Population aging</a><ul>
<li class="chapter" data-level="12.3.1" data-path="demographic-theory-and-covid-19.html"><a href="demographic-theory-and-covid-19.html#stable-theory"><i class="fa fa-check"></i><b>12.3.1</b> Stable theory</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="demographic-theory-and-covid-19.html"><a href="demographic-theory-and-covid-19.html#keyfitzs-entropy"><i class="fa fa-check"></i><b>12.4</b> Keyfitz’s entropy</a></li>
<li class="chapter" data-level="12.5" data-path="demographic-theory-and-covid-19.html"><a href="demographic-theory-and-covid-19.html#loss-of-person-years-remaining"><i class="fa fa-check"></i><b>12.5</b> Loss of person years remaining</a></li>
<li class="chapter" data-level="12.6" data-path="demographic-theory-and-covid-19.html"><a href="demographic-theory-and-covid-19.html#stationary-theory"><i class="fa fa-check"></i><b>12.6</b> Stationary theory</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Mathematical Demography</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="coalescent" class="section level1">
<h1><span class="header-section-number">Chapter 10</span> Coalescent</h1>
<div id="outline-9" class="section level2">
<h2><span class="header-section-number">10.1</span> Outline</h2>
<ul>
<li>Big picture: What is “coalescent theory”?</li>
<li>Time to (T)MRCA</li>
<li>Simulation: Inferring population size</li>
<li>An application of Coalescent Theory</li>
</ul>
</div>
<div id="the-coalescent-expectations-of-the-past" class="section level2">
<h2><span class="header-section-number">10.2</span> The Coalescent: Expectations of the Past</h2>
<ul>
<li>Coalescent theory is not a theory.</li>
<li>It’s a model for the probability of different histories.
<ul>
<li>Think about it as a model about how sampling can lead to a history of a common ancestor.</li>
</ul></li>
<li>“The” coalescent is a bit confusing.
<ul>
<li>We’re not inferring the actual history of common ancestry, just the probabilities</li>
</ul></li>
</ul>
<p>An actual “picture”
<img src="/hdir/0/fmenares/Book/bookdown-master/images/cutter_diagram.png" /></p>
<ul>
<li><p>Top panel is a Fisher-Wright instance, ordered so that lines don’t cross.</p></li>
<li><p>Haplotype is a sequence (we are diploids, each contributing 2 haplotypes).</p>
<ul>
<li>Let’s just think of each line as an individual, for now.</li>
<li>Such that each column is a different generation and the lines are the links across generations between dots. Each dot in the columns represents a “child” (or a parent to the next generation).</li>
</ul></li>
<li><p>The dark purple dots are 8 sampled individuals. We can find The Most Recent Common Ancestor (TMRCA) of sample. This is the red dot in the middle of the graph.</p></li>
<li><p>We can make inferences on the entire population based on the sample of individuals.</p>
<ul>
<li>Our sample <span class="math inline">\(\neq\)</span> even all <em>extant</em> descendants of the MRCA. In other words, the sample is not an exhaustive set of the survivors</li>
<li>Since some of the descendants of the pink lines don’t survive, then our sample <span class="math inline">\(\neq\)</span> all of the descendants of the MRCA.</li>
<li>If we chose two descendants at random, would we always get same MRCA?</li>
</ul></li>
<li><p>When we model coalescence we are thinking backwards in time.</p></li>
</ul>
</div>
<div id="our-first-question-when-was-mrca" class="section level2">
<h2><span class="header-section-number">10.3</span> Our first question: When was MRCA?</h2>
<ul>
<li><em>If we sample two individuals (today), how long ago was their MRCA?</em> Note that we are not looking for “who” the MRCA is but rather the number of generations ago that they occurred.
<ul>
<li>Our answer will be in terms of the probability of MRCA being 1 generation ago, 2 generations ago, etc.</li>
<li>We’ll assume Fisher-Wright: constant N, each gen randomly picks parents.</li>
<li>The answer is surprisingly simple.</li>
</ul></li>
</ul>
<p>Let’s assume we have <span class="math inline">\(N\)</span> lines in Fisher-Wright (Note: for now we are not using <span class="math inline">\(2N\)</span>.) As a reference, think about the population being the dots from the last column of the diagram from the last section. Then this sample is trying to look backwards <span class="math inline">\(n\)</span> generations until they find their MRCA.</p>
<ul>
<li><p>The chance that two sampled people have same parent is <span class="math inline">\(1/N\)</span>.</p>
<ul>
<li>Thus, the probability of coalescence in the first generation is: <span class="math inline">\(P(T_{MRCA} = 1) = 1/N\)</span>. Explicitly, <span class="math inline">\(T_{MRCA} = 1\)</span> means that the coalescence of lines occurs in the first generation.</li>
<li>What’s the probability that coalescence occurs in the second generation?</li>
</ul>
<p><span class="math display">\[P(T_{MRCA} = 2) = \left(1-\frac{1}{N}\right)\frac{1}{N} \]</span></p>
<ul>
<li><p>Here, <span class="math inline">\(\left(1- \frac{1}{N}\right)\)</span> is the probability of non coalescence in the first generation and the second term <span class="math inline">\(\frac{1}{N}\)</span> is the probability of coalescence in the <span class="math inline">\(2^{nd}\)</span> generation. Therefore the probability of coalescence in the second generation is subject to non coalescence in the first generation.</p></li>
<li><p>What is <span class="math inline">\(P(T_{MRCA} = n)\)</span>? Continuing with the logic from the previous answer we have:</p></li>
</ul>
<p><span class="math display">\[\begin{aligned}
P(T_{MRCA} = 3) &amp;= \left(1-\frac{1}{N}\right)^{2}\frac{1}{N}\\
P(T_{MRCA} = 4) &amp;= \left(1-\frac{1}{N}\right)^{3}\frac{1}{N}\\
\vdots \\
P(T_{MRCA} = n) &amp;= \left(1-\frac{1}{N}\right)^{n-1}\frac{1}{N}\\
\end{aligned}\]</span></p>
<p><em>With a reasonably big population and with a long time scale, we can think about the problem in continuous time</em></p>
<ul>
<li>Let the hazard of coalescence be <span class="math inline">\(c = 1/N\)</span>.</li>
<li>Probability of coalescence at time <span class="math inline">\(t\)</span>: <span class="math inline">\(\ell(t) h(t) = e^{-ct} c\)</span></li>
<li>The expected time of coalescence is analogous to the average age of death which is the life expectancy at birth.</li>
</ul>
<p><span class="math display">\[\begin{aligned}
e_0 &amp;=\frac{T_0}{l_0}\\
 &amp; = \frac{\sum_0^{\infty} \ell_x}{1} \\
 &amp; = \int_0^{\infty} \ell_x dx \\
 &amp; = \int_0^{\infty} e^{-cx}dx \\
 &amp; = (\frac{-1}{c})e^{-cx}|_0^{\infty}\\
 &amp; = (\frac{-1}{c})(0-1)\\
 &amp; = \frac{1}{c}
 \end{aligned}\]</span></p>
<ul>
<li>Since <span class="math inline">\(c = 1/N\)</span>, then the expected waiting time until coalescence is <span class="math inline">\(\frac{1}{c}= N\)</span></li>
<li>What is expected time of coalescence? Think life expectancy.</li>
<li><span class="math inline">\(E(T_{MRCA})\)</span> if two samples: <span class="math inline">\(1/c = 1/(1/N) = N\)</span></li>
<li>We can look at this through simulations. The functions below create Fisher-Wright processes and get the time to coalescence.</li>
</ul></li>
</ul>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb164-1"><a href="coalescent.html#cb164-1"></a><span class="co">## Our functions</span></span>
<span id="cb164-2"><a href="coalescent.html#cb164-2"></a></span>
<span id="cb164-3"><a href="coalescent.html#cb164-3"></a><span class="co">## (1) Fisher-Wright with Mutations &quot;2&quot;</span></span>
<span id="cb164-4"><a href="coalescent.html#cb164-4"></a><span class="co">## (builds on function we used last time, but also returns history of draws which we can use to trace TMRCA)</span></span>
<span id="cb164-5"><a href="coalescent.html#cb164-5"></a></span>
<span id="cb164-6"><a href="coalescent.html#cb164-6"></a>fwm2 &lt;-<span class="st"> </span><span class="cf">function</span>(N, n_gen, <span class="dt">mu =</span> <span class="dv">0</span>, <span class="dt">start_types =</span> <span class="kw">paste</span>(<span class="dv">1</span><span class="op">:</span>N))</span>
<span id="cb164-7"><a href="coalescent.html#cb164-7"></a>{</span>
<span id="cb164-8"><a href="coalescent.html#cb164-8"></a>    <span class="co">## position vector, for sampling</span></span>
<span id="cb164-9"><a href="coalescent.html#cb164-9"></a>    pos &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span>N</span>
<span id="cb164-10"><a href="coalescent.html#cb164-10"></a>    <span class="co">## matrices to keep book on each generation</span></span>
<span id="cb164-11"><a href="coalescent.html#cb164-11"></a>    <span class="co">## Types</span></span>
<span id="cb164-12"><a href="coalescent.html#cb164-12"></a>    A &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="ot">NA</span>, <span class="dt">nrow =</span> n_gen, <span class="dt">ncol =</span> N)</span>
<span id="cb164-13"><a href="coalescent.html#cb164-13"></a>    <span class="co">## Positions</span></span>
<span id="cb164-14"><a href="coalescent.html#cb164-14"></a>    P &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="ot">NA</span>, <span class="dt">nrow =</span> n_gen, <span class="dt">ncol =</span> N)</span>
<span id="cb164-15"><a href="coalescent.html#cb164-15"></a>    x &lt;-<span class="st"> </span>start_types</span>
<span id="cb164-16"><a href="coalescent.html#cb164-16"></a>    <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>n_gen)</span>
<span id="cb164-17"><a href="coalescent.html#cb164-17"></a>    {</span>
<span id="cb164-18"><a href="coalescent.html#cb164-18"></a>        <span class="co">## this code is a bit different from last time</span></span>
<span id="cb164-19"><a href="coalescent.html#cb164-19"></a>        <span class="co">## note: this is a bit confusing, maybe we should assign A[i+1,] &lt;- x at the end of this loop and</span></span>
<span id="cb164-20"><a href="coalescent.html#cb164-20"></a>        <span class="co">## only go up to n_gen -1</span></span>
<span id="cb164-21"><a href="coalescent.html#cb164-21"></a>        A[i,] &lt;-<span class="st"> </span>x  <span class="co">## types of last generation</span></span>
<span id="cb164-22"><a href="coalescent.html#cb164-22"></a>        positions &lt;-<span class="st"> </span><span class="kw">sample</span>(pos, <span class="dt">size =</span> N, <span class="dt">replace =</span> T) <span class="co">## we sample the parents 1:N</span></span>
<span id="cb164-23"><a href="coalescent.html#cb164-23"></a>        P[i,] &lt;-<span class="st"> </span>positions <span class="co">## keep book on positions</span></span>
<span id="cb164-24"><a href="coalescent.html#cb164-24"></a>        x &lt;-<span class="st"> </span>A[i,positions] <span class="co">## get types of next generation</span></span>
<span id="cb164-25"><a href="coalescent.html#cb164-25"></a>        x &lt;-<span class="st"> </span><span class="kw">mut</span>(x, mu) <span class="co">## mutation</span></span>
<span id="cb164-26"><a href="coalescent.html#cb164-26"></a>        <span class="co">## x</span></span>
<span id="cb164-27"><a href="coalescent.html#cb164-27"></a>    }</span>
<span id="cb164-28"><a href="coalescent.html#cb164-28"></a>    <span class="kw">return</span>(<span class="kw">list</span>(<span class="dt">A=</span>A, <span class="dt">P=</span>P)) <span class="co">## A= matrix of types, each line a generation; P= matrix of positions </span></span>
<span id="cb164-29"><a href="coalescent.html#cb164-29"></a>}</span>
<span id="cb164-30"><a href="coalescent.html#cb164-30"></a></span>
<span id="cb164-31"><a href="coalescent.html#cb164-31"></a>mut &lt;-<span class="st"> </span><span class="cf">function</span>(x, mu)</span>
<span id="cb164-32"><a href="coalescent.html#cb164-32"></a>{</span>
<span id="cb164-33"><a href="coalescent.html#cb164-33"></a>    <span class="co">## m, the individuals that mutate</span></span>
<span id="cb164-34"><a href="coalescent.html#cb164-34"></a>    m &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">rbinom</span>(<span class="kw">length</span>(x), <span class="dv">1</span>, mu) <span class="op">==</span><span class="st"> </span><span class="dv">1</span>)</span>
<span id="cb164-35"><a href="coalescent.html#cb164-35"></a>    <span class="cf">if</span> (<span class="kw">length</span>(m) <span class="op">==</span><span class="st"> </span><span class="dv">0</span>) <span class="co">## if no-one mutates</span></span>
<span id="cb164-36"><a href="coalescent.html#cb164-36"></a>        <span class="kw">return</span>(x)</span>
<span id="cb164-37"><a href="coalescent.html#cb164-37"></a>    <span class="co">## add a suffix to their ID, so it will be unique (infinite alleles)</span></span>
<span id="cb164-38"><a href="coalescent.html#cb164-38"></a>    suffix &lt;-<span class="st"> </span><span class="dv">10000</span><span class="op">*</span><span class="kw">round</span>(<span class="kw">runif</span>(<span class="kw">length</span>(m)),<span class="dv">4</span>)</span>
<span id="cb164-39"><a href="coalescent.html#cb164-39"></a>    x[m] &lt;-<span class="st"> </span><span class="kw">paste0</span>(x[m], <span class="st">&quot;.&quot;</span>, suffix)</span>
<span id="cb164-40"><a href="coalescent.html#cb164-40"></a>    x</span>
<span id="cb164-41"><a href="coalescent.html#cb164-41"></a>}</span>
<span id="cb164-42"><a href="coalescent.html#cb164-42"></a></span>
<span id="cb164-43"><a href="coalescent.html#cb164-43"></a></span>
<span id="cb164-44"><a href="coalescent.html#cb164-44"></a><span class="co">##       Time to TMRCA      </span></span>
<span id="cb164-45"><a href="coalescent.html#cb164-45"></a></span>
<span id="cb164-46"><a href="coalescent.html#cb164-46"></a>get.coalescence.time &lt;-<span class="st"> </span><span class="cf">function</span>(P, i,j, <span class="dt">verbose =</span> F)</span>
<span id="cb164-47"><a href="coalescent.html#cb164-47"></a>{</span>
<span id="cb164-48"><a href="coalescent.html#cb164-48"></a></span>
<span id="cb164-49"><a href="coalescent.html#cb164-49"></a>    ending_individuals =<span class="st"> </span><span class="kw">c</span>(i,j) <span class="co">## two individuals we&#39;re tracking back in time</span></span>
<span id="cb164-50"><a href="coalescent.html#cb164-50"></a>    parents &lt;-<span class="st"> </span>ending_individuals</span>
<span id="cb164-51"><a href="coalescent.html#cb164-51"></a>    k_ago =<span class="st"> </span><span class="dv">0</span></span>
<span id="cb164-52"><a href="coalescent.html#cb164-52"></a>    PP =<span class="st"> </span>P <span class="co">## for verbose, tracks coalescence</span></span>
<span id="cb164-53"><a href="coalescent.html#cb164-53"></a>    <span class="cf">for</span> (k <span class="cf">in</span> <span class="kw">nrow</span>(P)<span class="op">:</span><span class="dv">1</span>) <span class="co"># row P = generation until 1 (moving backwards in time)</span></span>
<span id="cb164-54"><a href="coalescent.html#cb164-54"></a>    {</span>
<span id="cb164-55"><a href="coalescent.html#cb164-55"></a>        k_ago =<span class="st"> </span>k_ago <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb164-56"><a href="coalescent.html#cb164-56"></a>        <span class="cf">if</span> (verbose <span class="op">==</span><span class="st"> </span><span class="ot">TRUE</span>)</span>
<span id="cb164-57"><a href="coalescent.html#cb164-57"></a>            <span class="kw">print</span>(parents)</span>
<span id="cb164-58"><a href="coalescent.html#cb164-58"></a>        parents &lt;-<span class="st"> </span>P[k, parents]</span>
<span id="cb164-59"><a href="coalescent.html#cb164-59"></a>        PP[k, parents] &lt;-<span class="st"> </span>P[k, parents]<span class="op">*</span><span class="dv">10</span> <span class="op">+</span><span class="st"> </span>P[k, parents] <span class="co"># way to tag the parents. PP = P when we are not following (backwards) the individuals i, j</span></span>
<span id="cb164-60"><a href="coalescent.html#cb164-60"></a>        <span class="cf">if</span> (parents[<span class="dv">1</span>] <span class="op">==</span><span class="st"> </span>parents[<span class="dv">2</span>]) <span class="co"># Stop looking backwards when both parents are the same</span></span>
<span id="cb164-61"><a href="coalescent.html#cb164-61"></a>        {</span>
<span id="cb164-62"><a href="coalescent.html#cb164-62"></a>            <span class="cf">if</span>(verbose)</span>
<span id="cb164-63"><a href="coalescent.html#cb164-63"></a>                <span class="kw">print</span>(parents)</span>
<span id="cb164-64"><a href="coalescent.html#cb164-64"></a>            <span class="kw">ifelse</span>(verbose,</span>
<span id="cb164-65"><a href="coalescent.html#cb164-65"></a>                   <span class="kw">return</span>(<span class="kw">list</span>(PP, k_ago)),</span>
<span id="cb164-66"><a href="coalescent.html#cb164-66"></a>                   <span class="kw">return</span>(k_ago))</span>
<span id="cb164-67"><a href="coalescent.html#cb164-67"></a>            }</span>
<span id="cb164-68"><a href="coalescent.html#cb164-68"></a></span>
<span id="cb164-69"><a href="coalescent.html#cb164-69"></a>    }</span>
<span id="cb164-70"><a href="coalescent.html#cb164-70"></a>    <span class="cf">if</span> (k_ago <span class="op">==</span><span class="st"> </span><span class="kw">nrow</span>(P))</span>
<span id="cb164-71"><a href="coalescent.html#cb164-71"></a>        k_ago &lt;-<span class="st"> </span><span class="ot">NA</span></span>
<span id="cb164-72"><a href="coalescent.html#cb164-72"></a>        <span class="kw">ifelse</span>(verbose,</span>
<span id="cb164-73"><a href="coalescent.html#cb164-73"></a>           <span class="kw">return</span>(<span class="kw">list</span>(PP, k_ago)),</span>
<span id="cb164-74"><a href="coalescent.html#cb164-74"></a>           <span class="kw">return</span>(k_ago))</span>
<span id="cb164-75"><a href="coalescent.html#cb164-75"></a>}</span></code></pre></div>
<ul>
<li>First, consider a FW process with 5 generations and 5 people. The first row of the matrix A is the starting population of people types 1 through 5. The next generations (in each row) are sampled from the previous generation with replacement. If one type is not sampled into the next generation then it is ‘’lost’’ and cannot appear in future generations.
** NOTE FOR JOSH:** I am really confused about the difference between the A and P matrix, so I don’t know if I am writing the correct explanation above and below. Could you please explain more how P and A work in the code below and in the context of the get.time.coalesce fn?</li>
</ul>
<div class="sourceCode" id="cb165"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb165-1"><a href="coalescent.html#cb165-1"></a><span class="kw">set.seed</span>(<span class="dv">10</span>)</span>
<span id="cb165-2"><a href="coalescent.html#cb165-2"></a>out &lt;-<span class="st"> </span><span class="kw">fwm2</span>(<span class="dv">5</span>,<span class="dv">5</span>,<span class="dv">0</span>) <span class="co"># N = 5, #generations = 5, mutation =0</span></span>
<span id="cb165-3"><a href="coalescent.html#cb165-3"></a>P &lt;-<span class="st"> </span>out<span class="op">$</span>P</span>
<span id="cb165-4"><a href="coalescent.html#cb165-4"></a>A &lt;-<span class="st"> </span>out<span class="op">$</span>A</span>
<span id="cb165-5"><a href="coalescent.html#cb165-5"></a><span class="kw">print</span>(A)</span></code></pre></div>
<pre><code>##      [,1] [,2] [,3] [,4] [,5]
## [1,] &quot;1&quot;  &quot;2&quot;  &quot;3&quot;  &quot;4&quot;  &quot;5&quot; 
## [2,] &quot;3&quot;  &quot;1&quot;  &quot;2&quot;  &quot;4&quot;  &quot;3&quot; 
## [3,] &quot;1&quot;  &quot;1&quot;  &quot;1&quot;  &quot;3&quot;  &quot;2&quot; 
## [4,] &quot;1&quot;  &quot;2&quot;  &quot;2&quot;  &quot;2&quot;  &quot;1&quot; 
## [5,] &quot;2&quot;  &quot;2&quot;  &quot;1&quot;  &quot;2&quot;  &quot;1&quot;</code></pre>
<div class="sourceCode" id="cb167"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb167-1"><a href="coalescent.html#cb167-1"></a><span class="kw">print</span>(P)</span></code></pre></div>
<pre><code>##      [,1] [,2] [,3] [,4] [,5]
## [1,]    3    1    2    4    3
## [2,]    2    2    2    5    3
## [3,]    2    5    5    5    1
## [4,]    4    2    5    2    5
## [5,]    2    1    2    2    1</code></pre>
<ul>
<li>Here we look at a function that obtains the time to coalescence given a FW process. The output contains (in order) the parents of the individuals tracked, the position matrix and the time to coalescence. For individuals 1 and 2 <span class="math inline">\(T_{MRCA}=3\)</span> and for individuals 3 and 4, <span class="math inline">\(T_{MRCA}=1\)</span></li>
</ul>
<div class="sourceCode" id="cb169"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb169-1"><a href="coalescent.html#cb169-1"></a><span class="kw">get.coalescence.time</span>(P,<span class="dv">1</span>,<span class="dv">2</span>, <span class="dt">verbose =</span> T) <span class="co"># individuals tracked are 1 and 2</span></span></code></pre></div>
<pre><code>## [1] 1 2
## [1] 2 1
## [1] 2 4
## [1] 5 5</code></pre>
<pre><code>## [[1]]
##      [,1] [,2] [,3] [,4] [,5]
## [1,]    3    1    2    4    3
## [2,]    2    2    2    5    3
## [3,]    2    5    5    5   11
## [4,]    4   22    5   22    5
## [5,]   22   11    2    2    1
## 
## [[2]]
## [1] 3</code></pre>
<div class="sourceCode" id="cb172"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb172-1"><a href="coalescent.html#cb172-1"></a><span class="kw">get.coalescence.time</span>(P,<span class="dv">3</span>,<span class="dv">4</span>, <span class="dt">verbose =</span> T) <span class="co"># individuals tracked are 3 and 4</span></span></code></pre></div>
<pre><code>## [1] 3 4
## [1] 2 2</code></pre>
<pre><code>## [[1]]
##      [,1] [,2] [,3] [,4] [,5]
## [1,]    3    1    2    4    3
## [2,]    2    2    2    5    3
## [3,]    2    5    5    5    1
## [4,]    4    2    5    2    5
## [5,]    2   11    2    2    1
## 
## [[2]]
## [1] 1</code></pre>
<div class="sourceCode" id="cb175"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb175-1"><a href="coalescent.html#cb175-1"></a><span class="co">#output: 1) parents , 2) PP matrix, 3) </span></span></code></pre></div>
<ul>
<li>Here is an example where there is no coalescence for individuals 3 and 4. In this case, we are looking at 5 types over 2 generations.</li>
</ul>
<div class="sourceCode" id="cb176"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb176-1"><a href="coalescent.html#cb176-1"></a><span class="kw">set.seed</span>(<span class="dv">10</span>)</span>
<span id="cb176-2"><a href="coalescent.html#cb176-2"></a>out &lt;-<span class="st"> </span><span class="kw">fwm2</span>(<span class="dv">5</span>,<span class="dv">2</span>,<span class="dv">0</span>)</span>
<span id="cb176-3"><a href="coalescent.html#cb176-3"></a>P &lt;-<span class="st"> </span>out<span class="op">$</span>P</span>
<span id="cb176-4"><a href="coalescent.html#cb176-4"></a><span class="kw">print</span>(P)</span></code></pre></div>
<pre><code>##      [,1] [,2] [,3] [,4] [,5]
## [1,]    3    1    2    4    3
## [2,]    2    2    2    5    3</code></pre>
<div class="sourceCode" id="cb178"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb178-1"><a href="coalescent.html#cb178-1"></a><span class="kw">get.coalescence.time</span>(P,<span class="dv">1</span>,<span class="dv">2</span>, <span class="dt">verbose =</span> T) <span class="co">#Individuals 1 and 2</span></span></code></pre></div>
<pre><code>## [1] 1 2
## [1] 2 2</code></pre>
<pre><code>## [[1]]
##      [,1] [,2] [,3] [,4] [,5]
## [1,]    3    1    2    4    3
## [2,]    2   22    2    5    3
## 
## [[2]]
## [1] 1</code></pre>
<div class="sourceCode" id="cb181"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb181-1"><a href="coalescent.html#cb181-1"></a><span class="kw">get.coalescence.time</span>(P,<span class="dv">3</span>,<span class="dv">4</span>, <span class="dt">verbose =</span> T) <span class="co">#Individuals 3 and 4</span></span></code></pre></div>
<pre><code>## [1] 3 4
## [1] 2 5</code></pre>
<pre><code>## [[1]]
##      [,1] [,2] [,3] [,4] [,5]
## [1,]   33    1   22    4    3
## [2,]    2   22    2    5   33
## 
## [[2]]
## [1] NA</code></pre>
<div class="sourceCode" id="cb184"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb184-1"><a href="coalescent.html#cb184-1"></a><span class="kw">get.coalescence.time</span>(P,<span class="dv">3</span>,<span class="dv">3</span>, <span class="dt">verbose =</span> T) <span class="co">#Individuals 3 and 3</span></span></code></pre></div>
<pre><code>## [1] 3 3
## [1] 2 2</code></pre>
<pre><code>## [[1]]
##      [,1] [,2] [,3] [,4] [,5]
## [1,]    3    1    2    4    3
## [2,]    2   22    2    5    3
## 
## [[2]]
## [1] 1</code></pre>
<ul>
<li><p>Now let’s do average of coalescent time in different FW processes. We’ll just draw 1 time for each one. We’ll look at 200 generations of 40 types (no mutations). Over 200 simulations, we find that the average time to coalescence is 52.08 which is close to what our formula <span class="math inline">\(E(T_{RMCA})=N\)</span> predicts. Try another size of N, ie 20, and you’ll find similar results!</p>
<ul>
<li>The variance is 2828.174. If the process was an exponential distribution we would get that the mean <span class="math inline">\(=\frac{1}{\lambda}\)</span> and that the variance <span class="math inline">\(=\frac{1}{\lambda^2}\)</span>. Here, if <span class="math inline">\(\lambda=\frac{1}{52.08}\)</span>, then then variance should be <span class="math inline">\(=\frac{1}{52.08^2}= 2712.326\)</span>. So, the variance below is a bit off from what we would expect from an exponential distribution.</li>
</ul></li>
</ul>
<div class="sourceCode" id="cb187"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb187-1"><a href="coalescent.html#cb187-1"></a><span class="kw">set.seed</span>(<span class="dv">2</span>)</span>
<span id="cb187-2"><a href="coalescent.html#cb187-2"></a>N &lt;-<span class="st"> </span><span class="dv">50</span> <span class="co">## pop size</span></span>
<span id="cb187-3"><a href="coalescent.html#cb187-3"></a>n_trials =<span class="st"> </span><span class="dv">200</span> <span class="co">## the number of &quot;histories&quot; we simulate</span></span>
<span id="cb187-4"><a href="coalescent.html#cb187-4"></a>n_samples =<span class="st"> </span><span class="dv">1</span> <span class="co">## the number of samples we do, of each history</span></span>
<span id="cb187-5"><a href="coalescent.html#cb187-5"></a>T.bar.vec &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="ot">NA</span>, n_trials)</span>
<span id="cb187-6"><a href="coalescent.html#cb187-6"></a><span class="cf">for</span> (r <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>n_trials) { <span class="co">## for each trial</span></span>
<span id="cb187-7"><a href="coalescent.html#cb187-7"></a>    <span class="co">#print(r)</span></span>
<span id="cb187-8"><a href="coalescent.html#cb187-8"></a>    out&lt;-<span class="st"> </span><span class="kw">fwm2</span>(N, <span class="dt">n_gen =</span> N<span class="op">*</span><span class="dv">20</span>, <span class="dt">mu =</span> <span class="dv">0</span>) <span class="co">## big enough n_gen so we almost always observe MRCA</span></span>
<span id="cb187-9"><a href="coalescent.html#cb187-9"></a>    P &lt;-<span class="st"> </span>out<span class="op">$</span>P</span>
<span id="cb187-10"><a href="coalescent.html#cb187-10"></a>    T.vec &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="ot">NA</span>, n_samples) <span class="co">## vector of Coalescent times from samples</span></span>
<span id="cb187-11"><a href="coalescent.html#cb187-11"></a>    <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>n_samples)</span>
<span id="cb187-12"><a href="coalescent.html#cb187-12"></a>    {</span>
<span id="cb187-13"><a href="coalescent.html#cb187-13"></a>        ij &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span>N, <span class="dv">2</span>) <span class="co">## draw random pair of individuals</span></span>
<span id="cb187-14"><a href="coalescent.html#cb187-14"></a>        T.vec[i] =<span class="st"> </span><span class="kw">get.coalescence.time</span>(P, ij[<span class="dv">1</span>], ij[<span class="dv">2</span>])</span>
<span id="cb187-15"><a href="coalescent.html#cb187-15"></a>    }</span>
<span id="cb187-16"><a href="coalescent.html#cb187-16"></a>    T.bar.vec[r] &lt;-<span class="st"> </span><span class="kw">mean</span>(T.vec) <span class="co">## average coalescent times for this trial</span></span>
<span id="cb187-17"><a href="coalescent.html#cb187-17"></a>}</span>
<span id="cb187-18"><a href="coalescent.html#cb187-18"></a></span>
<span id="cb187-19"><a href="coalescent.html#cb187-19"></a><span class="kw">mean</span>(T.bar.vec) <span class="co">#Mean time to coalescence</span></span></code></pre></div>
<pre><code>## [1] 52.08</code></pre>
<div class="sourceCode" id="cb189"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb189-1"><a href="coalescent.html#cb189-1"></a><span class="kw">var</span>(T.bar.vec) <span class="co">#Variance time to coalescence</span></span></code></pre></div>
<pre><code>## [1] 2828.174</code></pre>
<div class="sourceCode" id="cb191"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb191-1"><a href="coalescent.html#cb191-1"></a><span class="kw">set.seed</span>(<span class="dv">10</span>)</span>
<span id="cb191-2"><a href="coalescent.html#cb191-2"></a>out &lt;-<span class="st"> </span><span class="kw">fwm2</span>(<span class="dt">N=</span><span class="dv">5</span>,<span class="dt">n_gen=</span> <span class="dv">5</span>,<span class="dt">mu=</span><span class="dv">0</span>)</span>
<span id="cb191-3"><a href="coalescent.html#cb191-3"></a>P &lt;-<span class="st"> </span>out<span class="op">$</span>P</span>
<span id="cb191-4"><a href="coalescent.html#cb191-4"></a><span class="kw">print</span>(P)</span></code></pre></div>
<pre><code>##      [,1] [,2] [,3] [,4] [,5]
## [1,]    3    1    2    4    3
## [2,]    2    2    2    5    3
## [3,]    2    5    5    5    1
## [4,]    4    2    5    2    5
## [5,]    2    1    2    2    1</code></pre>
</div>
<div id="mutation-and-inference-of-tmrca-and-n" class="section level2">
<h2><span class="header-section-number">10.4</span> Mutation and inference of TMRCA and <span class="math inline">\(N\)</span></h2>
<ul>
<li>Using mutations, we can infer the most recent common ancestor. How?
<ul>
<li>Say mutations occur at a constant rate <span class="math inline">\(\mu\)</span> (<span class="math inline">\(10^{-8}\)</span>?). Then in each generation we would expect <span class="math inline">\(\mu\)</span> mutations, and over <span class="math inline">\(T\)</span> years we would expect <span class="math inline">\(T\mu\)</span> mutations. Specifically, since the same mutation cannot happen twice, each year we expect new mutations.</li>
<li>Say we observe that two people differ at <span class="math inline">\(k\)</span> sites of the genome. Given this information we can find the TMRCA and even the size of the population.</li>
</ul></li>
</ul>
<div id="htmlwidget-a3851c9c9a8f32e39c2b" style="width:200px;height:200px;" class="grViz html-widget"></div>
<script type="application/json" data-for="htmlwidget-a3851c9c9a8f32e39c2b">{"x":{"diagram":"digraph {\n\ngraph [layout = \"dot\",\n       rankdir = \"TB\",\n       outputorder = \"edgesfirst\",\n       bgcolor = \"white\"]\n\nnode [fontname = \"Helvetica\",\n      fontsize = \"10\",\n      shape = \"circle\",\n      fixedsize = \"true\",\n      width = \"0.5\",\n      style = \"filled\",\n      fillcolor = \"aliceblue\",\n      color = \"gray70\",\n      fontcolor = \"gray50\"]\n\nedge [fontname = \"Helvetica\",\n     fontsize = \"8\",\n     len = \"1.5\",\n     color = \"gray80\",\n     arrowsize = \"0.5\"]\n\n  \"1\" [fillcolor = \"#F0F8FF\", fontcolor = \"#000000\"] \n  \"2\" [fillcolor = \"#F0F8FF\", fontcolor = \"#000000\"] \n  \"3\" [fillcolor = \"#F0F8FF\", fontcolor = \"#000000\"] \n  \"1\"->\"2\" \n  \"1\"->\"3\" \n}","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script>
<p>In the tree graph above, we have that the difference between the parent point (1) and the branches is the number of generations:</p>
<ul>
<li><p>The tree length is <span class="math inline">\(2T\)</span></p>
<ul>
<li>The expected number of mutations is # branches <span class="math inline">\(\times\)</span> # generations <span class="math inline">\(\times\)</span> mutations:
<span class="math display">\[ E(k) = E(2T\mu) = 2\mu E(T) = 2 \mu \bar{T}  \]</span></li>
<li><span class="math inline">\(\bar{T}\)</span> is the average number of generations until we reach the most common recent ancestor and from the previous section we know that: <span class="math display">\[ \bar{T} = E(T_{MRCA}) = N \]</span></li>
<li>If we observe on average <span class="math inline">\(\bar{k}\)</span> mutations, then</li>
</ul>
<p><span class="math display">\[ \begin{aligned} 
\bar{k} &amp; \approx E(k) = N 2 \mu \\
\hat{N} &amp; = {\bar{k} \over 2 \mu}
\end{aligned} \]</span></p>
<ul>
<li>Therefore, an estimate of the population is given by the number of events (<span class="math inline">\(\bar{k}\)</span>) as a share of the exposure (<span class="math inline">\(2\mu\)</span>). This allows us to make inferences on how big a population is just by knowing an estimate of mutations and their probabilities.</li>
</ul></li>
</ul>
<div id="inference-of-population-size-simulation" class="section level3">
<h3><span class="header-section-number">10.4.1</span> Inference of population size (Simulation)</h3>
<ul>
<li>For this simulation, we will do FW with mutations. Then we will average pairwise differences and ivide by <span class="math inline">\(2\mu\)</span> to get our estimate.
<ul>
<li>We can repeat this process multiple times and see if the average estimate converges to the truth.</li>
</ul></li>
</ul>
<div class="sourceCode" id="cb193"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb193-1"><a href="coalescent.html#cb193-1"></a><span class="co"># Function</span></span>
<span id="cb193-2"><a href="coalescent.html#cb193-2"></a>count_pairwise_diff &lt;-<span class="st"> </span><span class="cf">function</span>(str1, str2, <span class="dt">verbose =</span> F)</span>
<span id="cb193-3"><a href="coalescent.html#cb193-3"></a>{</span>
<span id="cb193-4"><a href="coalescent.html#cb193-4"></a>    <span class="co">## stategy: split strings at &quot;.&quot; and add</span></span>
<span id="cb193-5"><a href="coalescent.html#cb193-5"></a>    <span class="co">## up number of mutations  after divergence</span></span>
<span id="cb193-6"><a href="coalescent.html#cb193-6"></a></span>
<span id="cb193-7"><a href="coalescent.html#cb193-7"></a>    <span class="cf">if</span> (verbose <span class="op">==</span><span class="st"> </span><span class="ot">TRUE</span>)</span>
<span id="cb193-8"><a href="coalescent.html#cb193-8"></a>        {</span>
<span id="cb193-9"><a href="coalescent.html#cb193-9"></a>            <span class="kw">print</span>(<span class="kw">paste0</span>(<span class="st">&#39;Item 1: &#39;</span>,str1))</span>
<span id="cb193-10"><a href="coalescent.html#cb193-10"></a>            <span class="kw">print</span>(<span class="kw">paste0</span>(<span class="st">&#39;Item 2: &#39;</span>,str2))</span>
<span id="cb193-11"><a href="coalescent.html#cb193-11"></a></span>
<span id="cb193-12"><a href="coalescent.html#cb193-12"></a>        }</span>
<span id="cb193-13"><a href="coalescent.html#cb193-13"></a>    z1 =<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">strsplit</span>(str1, <span class="st">&quot;</span><span class="ch">\\</span><span class="st">.&quot;</span>))</span>
<span id="cb193-14"><a href="coalescent.html#cb193-14"></a>    z2 =<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">strsplit</span>(str2, <span class="st">&quot;</span><span class="ch">\\</span><span class="st">.&quot;</span>))</span>
<span id="cb193-15"><a href="coalescent.html#cb193-15"></a>    n1 =<span class="st"> </span><span class="kw">length</span>(z1)</span>
<span id="cb193-16"><a href="coalescent.html#cb193-16"></a>    n2 =<span class="st"> </span><span class="kw">length</span>(z2)</span>
<span id="cb193-17"><a href="coalescent.html#cb193-17"></a>    n =<span class="st"> </span><span class="kw">min</span>(<span class="kw">c</span>(n1, n2))</span>
<span id="cb193-18"><a href="coalescent.html#cb193-18"></a>    <span class="co">## make sure population has fixed</span></span>
<span id="cb193-19"><a href="coalescent.html#cb193-19"></a>    <span class="cf">if</span>(z1[<span class="dv">1</span>] <span class="op">!=</span><span class="st"> </span>z2[<span class="dv">1</span>])</span>
<span id="cb193-20"><a href="coalescent.html#cb193-20"></a>        <span class="kw">return</span>(<span class="ot">NA</span>)</span>
<span id="cb193-21"><a href="coalescent.html#cb193-21"></a>    <span class="co">## now get location of prefix (ID of TMRCA) (ie last point before divergence)</span></span>
<span id="cb193-22"><a href="coalescent.html#cb193-22"></a>    pre.n =<span class="st"> </span><span class="kw">max</span>(<span class="kw">which</span>(z1[<span class="dv">1</span><span class="op">:</span>n] <span class="op">==</span><span class="st"> </span>z2[<span class="dv">1</span><span class="op">:</span>n]))</span>
<span id="cb193-23"><a href="coalescent.html#cb193-23"></a>    <span class="co">## get suffix</span></span>
<span id="cb193-24"><a href="coalescent.html#cb193-24"></a>    <span class="co">## these are the mutations after prefix</span></span>
<span id="cb193-25"><a href="coalescent.html#cb193-25"></a>    <span class="co">## code is a bit awkward because for 1 of the</span></span>
<span id="cb193-26"><a href="coalescent.html#cb193-26"></a>    <span class="co">## individuals, there may be no additional mutations</span></span>
<span id="cb193-27"><a href="coalescent.html#cb193-27"></a>    <span class="cf">if</span> (pre.n <span class="op">&lt;</span><span class="st"> </span><span class="kw">length</span>(z1))</span>
<span id="cb193-28"><a href="coalescent.html#cb193-28"></a>        post<span class="fl">.1</span> &lt;-<span class="st"> </span>z1[(pre.n <span class="op">+</span><span class="dv">1</span>)<span class="op">:</span><span class="st"> </span><span class="kw">length</span>(z1)]</span>
<span id="cb193-29"><a href="coalescent.html#cb193-29"></a>    <span class="cf">else</span></span>
<span id="cb193-30"><a href="coalescent.html#cb193-30"></a>        post<span class="fl">.1</span> =<span class="st"> </span><span class="ot">NULL</span></span>
<span id="cb193-31"><a href="coalescent.html#cb193-31"></a>    <span class="cf">if</span> (pre.n <span class="op">&lt;</span><span class="st"> </span><span class="kw">length</span>(z2))</span>
<span id="cb193-32"><a href="coalescent.html#cb193-32"></a>        post<span class="fl">.2</span> &lt;-<span class="st"> </span>z2[(pre.n <span class="op">+</span><span class="dv">1</span>)<span class="op">:</span><span class="st"> </span><span class="kw">length</span>(z2)]</span>
<span id="cb193-33"><a href="coalescent.html#cb193-33"></a>    <span class="cf">else</span></span>
<span id="cb193-34"><a href="coalescent.html#cb193-34"></a>        post<span class="fl">.2</span> =<span class="st"> </span><span class="ot">NULL</span></span>
<span id="cb193-35"><a href="coalescent.html#cb193-35"></a>    <span class="cf">if</span> (verbose <span class="op">==</span><span class="st"> </span><span class="ot">TRUE</span>) {</span>
<span id="cb193-36"><a href="coalescent.html#cb193-36"></a>        <span class="kw">print</span>(<span class="kw">paste0</span>(<span class="st">&#39;Mutations of item 1: &#39;</span>, <span class="kw">paste</span>(post<span class="fl">.1</span>, <span class="dt">collapse =</span> <span class="st">&quot;; &quot;</span>)))</span>
<span id="cb193-37"><a href="coalescent.html#cb193-37"></a>        <span class="kw">print</span>(<span class="kw">paste0</span>(<span class="st">&#39;Mutations of item 2: &#39;</span>, <span class="kw">paste</span>(post<span class="fl">.2</span>, <span class="dt">collapse =</span> <span class="st">&quot;; &quot;</span>)))</span>
<span id="cb193-38"><a href="coalescent.html#cb193-38"></a>    }</span>
<span id="cb193-39"><a href="coalescent.html#cb193-39"></a>    <span class="co">## count number of periods, note: length(NULL) is 0</span></span>
<span id="cb193-40"><a href="coalescent.html#cb193-40"></a>    n_mutations =<span class="st"> </span><span class="kw">length</span>(post<span class="fl">.1</span>) <span class="op">+</span><span class="st"> </span><span class="kw">length</span>(post<span class="fl">.2</span>)</span>
<span id="cb193-41"><a href="coalescent.html#cb193-41"></a></span>
<span id="cb193-42"><a href="coalescent.html#cb193-42"></a><span class="kw">return</span>(n_mutations)</span>
<span id="cb193-43"><a href="coalescent.html#cb193-43"></a>}</span></code></pre></div>
<ul>
<li>Let’s see how this function works by starting off with a FW process of 100 types in 2000 generations with a mutation rate <span class="math inline">\(\mu=0.01\)</span>. We save the last generation. Notice that each item in the last generation is a string with ‘.’ which denote a mutation and that the
they most of them only differ in the last 4 items between the dots.</li>
</ul>
<div class="sourceCode" id="cb194"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb194-1"><a href="coalescent.html#cb194-1"></a><span class="kw">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb194-2"><a href="coalescent.html#cb194-2"></a>mu =<span class="st"> </span><span class="fl">.01</span></span>
<span id="cb194-3"><a href="coalescent.html#cb194-3"></a>N =<span class="st"> </span><span class="dv">100</span></span>
<span id="cb194-4"><a href="coalescent.html#cb194-4"></a>A &lt;-<span class="st"> </span><span class="kw">fwm2</span>(N, <span class="dv">2000</span>, <span class="dt">mu =</span> mu)<span class="op">$</span>A</span>
<span id="cb194-5"><a href="coalescent.html#cb194-5"></a>x &lt;-<span class="st"> </span>A[<span class="dv">2000</span>,] <span class="co"># Obtain the last generation </span></span>
<span id="cb194-6"><a href="coalescent.html#cb194-6"></a>x[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</span></code></pre></div>
<pre><code>##  [1] &quot;74.5520.5563.3239.6558.7128.7639.9055.2605.7108.7876.5694.4778.248.6772.4397.8568.9300.2922.9156.6429.4644.6302.989&quot;                
##  [2] &quot;74.5520.5563.3239.6558.7128.7639.9055.2605.7108.7876.5694.4778.248.6772.4397.8568.9300.2922.9156.6429.4644.6302.4596.5931.1038.3855&quot;
##  [3] &quot;74.5520.5563.3239.6558.7128.7639.9055.2605.7108.7876.5694.4778.248.6772.4397.8568.9300.2922.9156.6429.4644.6302.989&quot;                
##  [4] &quot;74.5520.5563.3239.6558.7128.7639.9055.2605.7108.7876.5694.4778.248.6772.4397.8568.9300.2922.9156.6429.4644.6302.4596.5931.1038&quot;     
##  [5] &quot;74.5520.5563.3239.6558.7128.7639.9055.2605.7108.7876.5694.4778.248.6772.4397.8568.9300.2922.9156.6429.4644.6302.989&quot;                
##  [6] &quot;74.5520.5563.3239.6558.7128.7639.9055.2605.7108.7876.5694.4778.248.6772.4397.8568.9300.2922.9156.6429.4644.6302.4596.5931.1038&quot;     
##  [7] &quot;74.5520.5563.3239.6558.7128.7639.9055.2605.7108.7876.5694.4778.248.6772.4397.8568.9300.2922.9156.6429.4644.6302.4596.5931.1038&quot;     
##  [8] &quot;74.5520.5563.3239.6558.7128.7639.9055.2605.7108.7876.5694.4778.248.6772.4397.8568.9300.2922.9156.6429.4644.6302.4596.5931.1038&quot;     
##  [9] &quot;74.5520.5563.3239.6558.7128.7639.9055.2605.7108.7876.5694.4778.248.6772.4397.8568.9300.2922.9156.6429.4644.6302.989&quot;                
## [10] &quot;74.5520.5563.3239.6558.7128.7639.9055.2605.7108.7876.5694.4778.248.6772.4397.8568.9300.2922.9156.6429.4644.6302&quot;</code></pre>
<ul>
<li>Below we look at the differences between the different types. Function prints out the items, the mutations that occur after the types start diverging and the total number of mutations between each of the types. For types 1 and 4, there are 4 mutations while for types 1 and 10 there is only 1 mutations. Finally, for between types 10 and 13 there are only 3 differences.</li>
</ul>
<div class="sourceCode" id="cb196"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb196-1"><a href="coalescent.html#cb196-1"></a><span class="kw">count_pairwise_diff</span>(x[<span class="dv">1</span>], x[<span class="dv">4</span>], <span class="dt">verbose =</span> T)<span class="co">#Difference between types 1 and 4</span></span></code></pre></div>
<pre><code>## [1] &quot;Item 1: 74.5520.5563.3239.6558.7128.7639.9055.2605.7108.7876.5694.4778.248.6772.4397.8568.9300.2922.9156.6429.4644.6302.989&quot;
## [1] &quot;Item 2: 74.5520.5563.3239.6558.7128.7639.9055.2605.7108.7876.5694.4778.248.6772.4397.8568.9300.2922.9156.6429.4644.6302.4596.5931.1038&quot;
## [1] &quot;Mutations of item 1: 989&quot;
## [1] &quot;Mutations of item 2: 4596; 5931; 1038&quot;</code></pre>
<pre><code>## [1] 4</code></pre>
<div class="sourceCode" id="cb199"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb199-1"><a href="coalescent.html#cb199-1"></a><span class="kw">count_pairwise_diff</span>(x[<span class="dv">1</span>], x[<span class="dv">10</span>], <span class="dt">verbose =</span> T)<span class="co">#Difference between types 1 and 10</span></span></code></pre></div>
<pre><code>## [1] &quot;Item 1: 74.5520.5563.3239.6558.7128.7639.9055.2605.7108.7876.5694.4778.248.6772.4397.8568.9300.2922.9156.6429.4644.6302.989&quot;
## [1] &quot;Item 2: 74.5520.5563.3239.6558.7128.7639.9055.2605.7108.7876.5694.4778.248.6772.4397.8568.9300.2922.9156.6429.4644.6302&quot;
## [1] &quot;Mutations of item 1: 989&quot;
## [1] &quot;Mutations of item 2: &quot;</code></pre>
<pre><code>## [1] 1</code></pre>
<div class="sourceCode" id="cb202"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb202-1"><a href="coalescent.html#cb202-1"></a><span class="kw">count_pairwise_diff</span>(x[<span class="dv">10</span>], x[<span class="dv">13</span>], <span class="dt">verbose =</span> T)<span class="co">#Difference between types 10 and 13</span></span></code></pre></div>
<pre><code>## [1] &quot;Item 1: 74.5520.5563.3239.6558.7128.7639.9055.2605.7108.7876.5694.4778.248.6772.4397.8568.9300.2922.9156.6429.4644.6302&quot;
## [1] &quot;Item 2: 74.5520.5563.3239.6558.7128.7639.9055.2605.7108.7876.5694.4778.248.6772.4397.8568.9300.2922.9156.6429.4644.6302.4596.5931.1038&quot;
## [1] &quot;Mutations of item 1: &quot;
## [1] &quot;Mutations of item 2: 4596; 5931; 1038&quot;</code></pre>
<pre><code>## [1] 3</code></pre>
<div class="sourceCode" id="cb205"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb205-1"><a href="coalescent.html#cb205-1"></a><span class="co">## now do get average of all pairwise differences</span></span>
<span id="cb205-2"><a href="coalescent.html#cb205-2"></a>x &lt;-<span class="st"> </span>A[<span class="dv">2000</span>,]</span>
<span id="cb205-3"><a href="coalescent.html#cb205-3"></a>Pij &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="ot">NA</span>, <span class="kw">length</span>(x), <span class="kw">length</span>(x))</span>
<span id="cb205-4"><a href="coalescent.html#cb205-4"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>(<span class="kw">length</span>(x)<span class="op">-</span><span class="dv">1</span>))</span>
<span id="cb205-5"><a href="coalescent.html#cb205-5"></a>{</span>
<span id="cb205-6"><a href="coalescent.html#cb205-6"></a>    <span class="cf">for</span> (j <span class="cf">in</span> (i<span class="op">+</span><span class="dv">1</span>)<span class="op">:</span><span class="kw">length</span>(x))</span>
<span id="cb205-7"><a href="coalescent.html#cb205-7"></a>        {</span>
<span id="cb205-8"><a href="coalescent.html#cb205-8"></a>            Pij[i,j] =<span class="st"> </span><span class="kw">count_pairwise_diff</span>(x[i], x[j]) <span class="co"># if verbose=F, then only prints out the total number of mutations</span></span>
<span id="cb205-9"><a href="coalescent.html#cb205-9"></a>        }</span>
<span id="cb205-10"><a href="coalescent.html#cb205-10"></a>}</span>
<span id="cb205-11"><a href="coalescent.html#cb205-11"></a></span>
<span id="cb205-12"><a href="coalescent.html#cb205-12"></a>k.bar =<span class="st"> </span><span class="kw">mean</span>(Pij, <span class="dt">na.rm =</span> T) <span class="co"># </span></span>
<span id="cb205-13"><a href="coalescent.html#cb205-13"></a></span>
<span id="cb205-14"><a href="coalescent.html#cb205-14"></a><span class="co">## we can now infer pop size</span></span>
<span id="cb205-15"><a href="coalescent.html#cb205-15"></a>N.hat =<span class="st"> </span>k.bar <span class="op">/</span>(<span class="dv">2</span><span class="op">*</span>mu)</span>
<span id="cb205-16"><a href="coalescent.html#cb205-16"></a><span class="kw">print</span>(N.hat)</span></code></pre></div>
<pre><code>## [1] 112.7778</code></pre>
<ul>
<li>Now do repeated trials to see if estimator approximates the formula from the previous section:
<span class="math inline">\(\hat{N} = \frac{\bar{k}}{2 \mu}\)</span>. With 100 simulations, we get really close to the actual size of the population, which we established at the beginning.</li>
</ul>
<div class="sourceCode" id="cb207"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb207-1"><a href="coalescent.html#cb207-1"></a>mu =<span class="st"> </span><span class="fl">.01</span></span>
<span id="cb207-2"><a href="coalescent.html#cb207-2"></a>n_trials =<span class="st"> </span><span class="dv">100</span></span>
<span id="cb207-3"><a href="coalescent.html#cb207-3"></a>N =<span class="st"> </span><span class="dv">100</span></span>
<span id="cb207-4"><a href="coalescent.html#cb207-4"></a>n_gen =<span class="st"> </span>N<span class="op">*</span><span class="dv">5</span></span>
<span id="cb207-5"><a href="coalescent.html#cb207-5"></a>k.bar.vec =<span class="st"> </span><span class="ot">NULL</span></span>
<span id="cb207-6"><a href="coalescent.html#cb207-6"></a>n_casesfixation =<span class="st"> </span><span class="dv">0</span> </span>
<span id="cb207-7"><a href="coalescent.html#cb207-7"></a><span class="cf">for</span> (r <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>n_trials)</span>
<span id="cb207-8"><a href="coalescent.html#cb207-8"></a>{</span>
<span id="cb207-9"><a href="coalescent.html#cb207-9"></a>    <span class="co">#print(r)</span></span>
<span id="cb207-10"><a href="coalescent.html#cb207-10"></a>    A &lt;-<span class="st"> </span><span class="kw">fwm2</span>(N, n_gen, <span class="dt">mu =</span> mu)<span class="op">$</span>A</span>
<span id="cb207-11"><a href="coalescent.html#cb207-11"></a>    x &lt;-<span class="st"> </span>A[n_gen,]</span>
<span id="cb207-12"><a href="coalescent.html#cb207-12"></a>    <span class="co">## check on fixation (if we have many of these, increase n_gen)</span></span>
<span id="cb207-13"><a href="coalescent.html#cb207-13"></a>    <span class="cf">if</span> (<span class="kw">length</span>(<span class="kw">unique</span>(<span class="kw">substr</span>(x,<span class="dv">1</span>,<span class="dv">2</span>))) <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>)</span>
<span id="cb207-14"><a href="coalescent.html#cb207-14"></a>        <span class="co">#print(x)</span></span>
<span id="cb207-15"><a href="coalescent.html#cb207-15"></a>        n_casesfixation =<span class="st"> </span>n_casesfixation <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb207-16"><a href="coalescent.html#cb207-16"></a>    Pij &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="ot">NA</span>, <span class="kw">length</span>(x), <span class="kw">length</span>(x))</span>
<span id="cb207-17"><a href="coalescent.html#cb207-17"></a>    <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>(<span class="kw">length</span>(x)<span class="op">-</span><span class="dv">1</span>))</span>
<span id="cb207-18"><a href="coalescent.html#cb207-18"></a>    {</span>
<span id="cb207-19"><a href="coalescent.html#cb207-19"></a>        <span class="cf">for</span> (j <span class="cf">in</span> (i<span class="op">+</span><span class="dv">1</span>)<span class="op">:</span><span class="kw">length</span>(x))</span>
<span id="cb207-20"><a href="coalescent.html#cb207-20"></a>        {</span>
<span id="cb207-21"><a href="coalescent.html#cb207-21"></a>            Pij[i,j] =<span class="st"> </span><span class="kw">count_pairwise_diff</span>(x[i], x[j])</span>
<span id="cb207-22"><a href="coalescent.html#cb207-22"></a>        }</span>
<span id="cb207-23"><a href="coalescent.html#cb207-23"></a>    }</span>
<span id="cb207-24"><a href="coalescent.html#cb207-24"></a>   k.bar.vec[r] &lt;-<span class="st"> </span><span class="kw">mean</span>(Pij, <span class="dt">na.rm =</span> T)</span>
<span id="cb207-25"><a href="coalescent.html#cb207-25"></a>}</span>
<span id="cb207-26"><a href="coalescent.html#cb207-26"></a><span class="kw">print</span>(<span class="kw">paste0</span>(<span class="st">&#39;# cases of fixation: &#39;</span>,n_casesfixation))</span></code></pre></div>
<pre><code>## [1] &quot;# cases of fixation: 4&quot;</code></pre>
<div class="sourceCode" id="cb209"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb209-1"><a href="coalescent.html#cb209-1"></a><span class="kw">print</span>(<span class="kw">paste0</span>(<span class="st">&#39;Average # mutations: &#39;</span>,<span class="kw">mean</span>(k.bar.vec)))</span></code></pre></div>
<pre><code>## [1] &quot;Average # mutations: 1.85187558619893&quot;</code></pre>
<div class="sourceCode" id="cb211"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb211-1"><a href="coalescent.html#cb211-1"></a>N.hat =<span class="st"> </span><span class="kw">mean</span>(k.bar.vec) <span class="op">/</span>(<span class="dv">2</span><span class="op">*</span>mu)</span>
<span id="cb211-2"><a href="coalescent.html#cb211-2"></a><span class="kw">print</span>(<span class="kw">paste0</span>(<span class="st">&#39;Average population size: &#39;</span>,N.hat))</span></code></pre></div>
<pre><code>## [1] &quot;Average population size: 92.5937793099466&quot;</code></pre>
<ul>
<li>Given this simulation framework, we can modify many of the inputs to see how the size of the population reacts. For instance, we can include a large mutation rate or modify the population size.</li>
</ul>
</div>
</div>
<div id="coalescence-of-a-sample-of-n-individuals" class="section level2">
<h2><span class="header-section-number">10.5</span> Coalescence of a sample of <span class="math inline">\(n\)</span> individuals</h2>
<ul>
<li>This is covered on pages 42 and 43 of Gillespie
<strong>NOTE FOR JOSH: I can’t find this reference. What are you refering to?</strong></li>
<li>We’ll just do one quick example, accepting the result</li>
<li>Here is a sample of 3 (Note we’re using <span class="math inline">\(N\)</span> instead of <span class="math inline">\(2N\)</span>).</li>
<li>The intuition here is that when we have more individuals, there’s more chance that some pair of them will coalesce.</li>
<li>This tree might be the beginning or the end, we really don’t know. So, it is important to keep in mind what we are talking about.</li>
</ul>
<div class="sourceCode" id="cb213"><pre class="sourceCode latex"><code class="sourceCode latex"><span id="cb213-1"><a href="coalescent.html#cb213-1"></a><span class="kw">\begin</span>{<span class="ex">verbatim</span>}</span>
<span id="cb213-2"><a href="coalescent.html#cb213-2"></a></span>
<span id="cb213-3"><a href="coalescent.html#cb213-3"></a><span class="vs">              *            _______</span></span>
<span id="cb213-4"><a href="coalescent.html#cb213-4"></a><span class="vs">             / \</span></span>
<span id="cb213-5"><a href="coalescent.html#cb213-5"></a><span class="vs">            /   \          </span></span>
<span id="cb213-6"><a href="coalescent.html#cb213-6"></a><span class="vs">           /     \         T(2) : E(T(2)) = N</span></span>
<span id="cb213-7"><a href="coalescent.html#cb213-7"></a><span class="vs">          /       \</span></span>
<span id="cb213-8"><a href="coalescent.html#cb213-8"></a><span class="vs">         /         *       _______</span></span>
<span id="cb213-9"><a href="coalescent.html#cb213-9"></a><span class="vs">        /         / \</span></span>
<span id="cb213-10"><a href="coalescent.html#cb213-10"></a><span class="vs">       /         /   \     T(3) : E(T(3)) = N * 2/[3*2] = N / 3</span></span>
<span id="cb213-11"><a href="coalescent.html#cb213-11"></a><span class="vs">      /         /     *    _______</span></span>
<span id="cb213-12"><a href="coalescent.html#cb213-12"></a><span class="vs">     /         /     / \</span></span>
<span id="cb213-13"><a href="coalescent.html#cb213-13"></a><span class="vs">    /         /     /   \  T(4) : E(T(4)) = N * 2/[4*3] = N / 6</span></span>
<span id="cb213-14"><a href="coalescent.html#cb213-14"></a><span class="vs">   *         *      *    *</span></span>
<span id="cb213-15"><a href="coalescent.html#cb213-15"></a><span class="vs">   </span></span>
<span id="cb213-16"><a href="coalescent.html#cb213-16"></a><span class="vs">                            ...</span></span>
<span id="cb213-17"><a href="coalescent.html#cb213-17"></a><span class="vs">                           _______</span></span>
<span id="cb213-18"><a href="coalescent.html#cb213-18"></a><span class="vs">                           T(n) : E(T(n)) = N * 2/[n * (n-1)]</span></span>
<span id="cb213-19"><a href="coalescent.html#cb213-19"></a><span class="kw">\end</span>{<span class="ex">verbatim</span>}</span></code></pre></div>
<ul>
<li>If we sample 4, how much of time to TMRCA is do we for 4 branches, 3 branches, and 2 branches?
<strong>NOTE FOR JOSH: Could you check this answer here</strong>
<ul>
<li>The total time waiting until MRCA of all 4 individuals is <span class="math inline">\(N + {N\over 3} + {N \over 6} = {3\over 2} N\)</span></li>
<li>Then, the time spent at each branch relative to the total time (out of 4 branches) is given by the third column.</li>
<li>We count the total branch length as the number of branches time the expected time to coalesce.</li>
</ul></li>
</ul>
<p>|———————-|————-|————————————————–|——————————|
| Number of branches (n) | E(T(n)) | Time spent at each branch /total | Total branch length |
|————————|————-|————————————————–|——————————|
| 2 | <span class="math inline">\(N\)</span> | <span class="math inline">\(\frac{N}{N\frac{3}{2}}=\frac{2}{3}\)</span> | <span class="math inline">\(2N\)</span> |
| 3 |<span class="math inline">\(\frac{N}{3}\)</span>| <span class="math inline">\(\frac{N\frac{1}{3}}{N\frac{3}{2}}= \frac{2}{9}\)</span> |<span class="math inline">\(3\frac{N}{3}=N\)</span> |
| 4 |<span class="math inline">\(\frac{N}{6}\)</span>| <span class="math inline">\(\frac{N\frac{1}{6}}{N\frac{3}{2}}= \frac{1}{9}\)</span> |<span class="math inline">\(4\frac{N}{6} = \frac{2}{3}N\)</span> |</p>
</div>
<div id="an-application-of-coalescent-theory" class="section level2">
<h2><span class="header-section-number">10.6</span> An application of coalescent theory</h2>
<p>In this section, we will look at an application to making inferences about the real history of human populations. <em>Caveat: We’re still using <span class="math inline">\(N\)</span> (not <span class="math inline">\(2N\)</span>) as the number of haploids</em></p>
<div id="coalesence-when-population-is-changing" class="section level3">
<h3><span class="header-section-number">10.6.1</span> Coalesence when population is changing</h3>
<ul>
<li>We know that the hazard of coalescence is <span class="math inline">\(h = c = 1/N\)</span>. For instance if the populations were <span class="math inline">\(N_1 =1000\)</span>, <span class="math inline">\(N_2 = 2000\)</span>, then the hazard of coalesence in one generation for each of the populations would be <span class="math inline">\(h_1= {1\over1000}\)</span> and <span class="math inline">\(h_2= {1\over2000}\)</span>, respectively.</li>
<li>What if within the same population in one generation we have <span class="math inline">\(N(t) = 1000\)</span> and in the following generation <span class="math inline">\(N(t+1)= 2000\)</span>? We still follow FW in allowing children to choose their parents.</li>
<li>If the population size changes over time <span class="math inline">\(N(t)\)</span>, then hazards of coalescence in will change too: <span class="math inline">\(h(t) = 1/N(t)\)</span>.</li>
</ul>
<p>Let’s look at a small simulation. We are thinking about a population that has lived over a certain number of years: 5000 people lived during the last 1000 years. Before that, the population was only 500.
<strong>NOTE FOR JOSH: It’s not so clear to me why T1 is being sampled with N_recent and not with N_ancient. I thought that for the first 1000 years, the time to coalesce should be drawn from rexp(n, rate = 1/N_ancient) and then T2 &lt;- T_thresh + rexp(n2, rate = 1/N_recent). I tried changing in the simulation and I don’t get the nice histogram (nor the intended message), so I assume that my intuition is wrong. Could you clarify the populations involved and when the threshold happens in relation to the population sizes? </strong></p>
<div class="sourceCode" id="cb214"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb214-1"><a href="coalescent.html#cb214-1"></a>  N_recent =<span class="st"> </span><span class="dv">5000</span> <span class="co">## population last T_thresh years</span></span>
<span id="cb214-2"><a href="coalescent.html#cb214-2"></a>  T_thresh =<span class="st"> </span><span class="dv">1000</span></span>
<span id="cb214-3"><a href="coalescent.html#cb214-3"></a>  N_ancient =<span class="st"> </span><span class="dv">500</span> <span class="co">## earlier population</span></span>
<span id="cb214-4"><a href="coalescent.html#cb214-4"></a>  n =<span class="st"> </span><span class="dv">1000</span> <span class="co">## sampled individuals</span></span>
<span id="cb214-5"><a href="coalescent.html#cb214-5"></a>  <span class="kw">set.seed</span>(<span class="fl">0.4886</span>)</span>
<span id="cb214-6"><a href="coalescent.html#cb214-6"></a>  T1 &lt;-<span class="st"> </span><span class="kw">rexp</span>(n, <span class="dt">rate =</span> <span class="dv">1</span><span class="op">/</span>N_recent) <span class="co">## random draws from an exponential fn of times to coalesce. </span></span>
<span id="cb214-7"><a href="coalescent.html#cb214-7"></a>  T1[T1 <span class="op">&gt;</span><span class="st"> </span>T_thresh] &lt;-<span class="st"> </span><span class="ot">NA</span> <span class="co">## Remove times that coalesce event happens after first 1000 generations. </span></span>
<span id="cb214-8"><a href="coalescent.html#cb214-8"></a>  n2 &lt;-<span class="st"> </span><span class="kw">sum</span>(<span class="kw">is.na</span>(T1)) <span class="co"># total years removed. </span></span>
<span id="cb214-9"><a href="coalescent.html#cb214-9"></a>  T2 &lt;-<span class="st"> </span>T_thresh <span class="op">+</span><span class="st"> </span><span class="kw">rexp</span>(n2, <span class="dt">rate =</span> <span class="dv">1</span><span class="op">/</span>N_ancient) <span class="co">## at ancient rate</span></span>
<span id="cb214-10"><a href="coalescent.html#cb214-10"></a>  T.vec &lt;-<span class="st"> </span><span class="kw">c</span>(T1, T2)</span>
<span id="cb214-11"><a href="coalescent.html#cb214-11"></a>  <span class="kw">hist</span>(T.vec, <span class="dt">breaks =</span> <span class="kw">seq</span>(<span class="dv">0</span>, <span class="dv">5000</span>, <span class="dv">250</span>), <span class="dt">main=</span><span class="st">&quot;History of times to coalesce&quot;</span>, <span class="dt">xlab=</span><span class="st">&quot;Years&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-166"></span>
<img src="bookdown-demo_files/figure-html/unnamed-chunk-166-1.png" alt="Histogram" width="672" />
<p class="caption">
Figure 10.1: Histogram
</p>
</div>
<ul>
<li><em>How could we estimate population sizes from this histogram?</em> We need to go from hazard to population size. First, let’s think about the times to coalesce like the age at death of a specific line of types. Therefore, we sort the times to coalesce (equivalent to time of death) to get a survivorship graph and then obtain <span class="math inline">\(d_x\)</span>.</li>
</ul>
<div class="sourceCode" id="cb215"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb215-1"><a href="coalescent.html#cb215-1"></a>  T.vec &lt;-<span class="st"> </span><span class="kw">sort</span>(T.vec)</span>
<span id="cb215-2"><a href="coalescent.html#cb215-2"></a>  St =<span class="st"> </span>(n<span class="op">:</span><span class="dv">1</span>)<span class="op">/</span>n</span>
<span id="cb215-3"><a href="coalescent.html#cb215-3"></a>  <span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb215-4"><a href="coalescent.html#cb215-4"></a>  <span class="kw">plot</span>(T.vec, St)</span>
<span id="cb215-5"><a href="coalescent.html#cb215-5"></a>  <span class="kw">abline</span>(<span class="dt">v =</span> T_thresh)</span>
<span id="cb215-6"><a href="coalescent.html#cb215-6"></a>  <span class="kw">plot</span>(T.vec, <span class="kw">log</span>(St))</span>
<span id="cb215-7"><a href="coalescent.html#cb215-7"></a>  <span class="kw">abline</span>(<span class="dt">v =</span> T_thresh)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-167"></span>
<img src="bookdown-demo_files/figure-html/unnamed-chunk-167-1.png" alt="Population" width="672" />
<p class="caption">
Figure 10.2: Population
</p>
</div>
<ul>
<li><em>How can we estimate hazards from this histogram?</em>
<ul>
<li>Say we have <span class="math inline">\(i\)</span> pairs of haploids</li>
<li>We then compute how many pairwise differences there are, but instead of computing <span class="math inline">\(\bar{k}\)</span>, we keep the distributional information <span class="math inline">\(k_i\)</span>.</li>
<li>Each <span class="math inline">\(k_i\)</span> implies a <span class="math inline">\(T_i\)</span></li>
<li>We then have a set of ``death times’’ (coalescence times), can build a life table, estimate the hazards, and infer <span class="math inline">\(N(t)\)</span>.</li>
<li>Below we proceed with an example of how to obtain the hazards and the population sizes.</li>
</ul></li>
</ul>
</div>
</div>
<div id="reconstruction-ancient-european-population-sizes-using-batinis-sample-of-mitochondrial-dna" class="section level2">
<h2><span class="header-section-number">10.7</span> Reconstruction Ancient European Population Sizes using Batini’s sample of Mitochondrial DNA"</h2>
<ul>
<li><p>We use real sequences of mitochondrial DNA to estimate ancient population sizes. The sequences were made available by <span class="citation">Batini et al. (<a href="#ref-batini2017population" role="doc-biblioref">2017</a>)</span>, who analyzed sub-populations in their paper using software for Bayesian inference for comparing groups of individuals. Our approach here is to use a simpler, but less powerful approach. We will simply look at pairwise differences of random pairs of individuals.
<strong>NOTE FOR JOSH: Is the Batini reference this one? <a href="https://www.nature.com/articles/s41598-017-11307-9" class="uri">https://www.nature.com/articles/s41598-017-11307-9</a></strong></p>
<ul>
<li>For this we are not going to do subgroups such as the Greeks or Irish, but are going to use the entire European sample. The population sizes we estimate are for the entire population represented by the 328 individuals.</li>
</ul></li>
</ul>
<blockquote>
<p>Special thanks to Ken Wachter for this approach and whose R-code forms the basis of this application</p>
</blockquote>
<div id="summary-of-batini-et-al." class="section level3">
<h3><span class="header-section-number">10.7.1</span> Summary of Batini et al.</h3>
<p>Before we begin our own analysis, let’s look at the inputs to <span class="citation">Batini et al. (<a href="#ref-batini2017population" role="doc-biblioref">2017</a>)</span> analysis – the mtDNA haplotype sequences, and then at the resulting population estimates.</p>
<div id="the-mitochondrial-dna" class="section level4">
<h4><span class="header-section-number">10.7.1.1</span> The mitochondrial DNA</h4>
<p>Data preparation begins.</p>
<p>There are 380 individuals grouped into regional sub-populations. We select out the Greeks, whose labels begin with “gre”.</p>
<div class="sourceCode" id="cb216"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb216-1"><a href="coalescent.html#cb216-1"></a>x &lt;-<span class="st"> </span><span class="kw">scan</span>(<span class="st">&quot;/hdir/0/fmenares/Book/bookdown-master/data/mtdna.csv&quot;</span>, <span class="dt">what =</span> <span class="kw">character</span>())</span>
<span id="cb216-2"><a href="coalescent.html#cb216-2"></a><span class="kw">nchar</span>(x)</span></code></pre></div>
<pre><code>##   [1]     4     3     4     2     5    11     1    15 16577 16577 16577 16577 16577
##  [14] 16577 16577 16577 16577 16577 16577 16577 16577 16577 16577 16577 16577 16577
##  [27] 16574 16575 16575 16575 16575 16575 16575 16575 16575 16575 16575 16574 16575
##  [40] 16574 16574 16574 16574 16574 16574 16574 16575 16575 16575 16575 16575 16575
##  [53] 16575 16575 16575 16575 16575 16575 16575 16575 16575 16575 16575 16575 16575
##  [66] 16574 16576 16576 16576 16576 16576 16576 16576 16576 16576 16576 16576 16576
##  [79] 16576 16576 16575 16575 16575 16578 16578 16575 16575 16575 16576 16577 16577
##  [92] 16577 16577 16577 16577 16577 16577 16577 16577 16577 16577 16577 16577 16577
## [105] 16577 16577 16577 16577 16577 16575 16574 16575 16575 16575 16575 16575 16574
## [118] 16574 16574 16574 16575 16575 16575 16575 16575 16576 16576 16576 16575 16575
## [131] 16576 16576 16574 16574 16575 16575 16575 16575 16575 16574 16575 16575 16575
## [144] 16575 16575 16575 16575 16575 16575 16575 16574 16574 16575 16575 16575 16575
## [157] 16575 16575 16574 16575 16575 16575 16575 16575 16575 16575 16575 16575 16574
## [170] 16574 16574 16574 16580 16580 16580 16580 16580 16580 16580 16580 16580 16580
## [183] 16580 16580 16580 16580 16580 16580 16580 16580 16580 16580 16580 16580 16580
## [196] 16580 16580 16580 16580 16580 16580 16580 16580 16580 16580 16580 16580 16580
## [209] 16580 16580 16580 16580 16580 16580 16580 16580 16580 16580 16580 16580 16580
## [222] 16580 16580 16580 16580 16580 16580 16580 16580 16580 16580 16580 16580 16580
## [235] 16580 16580 16580 16580 16580 16580 16580 16580 16580 16580 16580 16580 16580
## [248] 16580 16580 16580 16580 16580 16575 16576 16576 16576 16576 16577 16577 16577
## [261] 16576 16576 16576 16576 16576 16576 16576 16576 16576 16576 16576 16576 16576
## [274] 16576 16576 16576 16576 16576 16576 16576 16576 16577 16577 16577 16577 16577
## [287] 16577 16577 16577 16577 16577 16577 16577 16577 16577 16576 16577 16577 16576
## [300] 16575 16577 16575 16575 16577 16574 16575 16575 16575 16575 16574 16575 16575
## [313] 16575 16575 16575 16575 16575 16574 16575 16574 16574 16574 16574 16574 16575
## [326] 16575 16574 16575 16575 16575 16576 16575 16575 16575 16575 16575 16575 16575
## [339] 16575 16576 16575 16576 16575 16575 16574 16575 16575 16575 16575 16575 16575
## [352] 16575 16575 16575 16575 16574 16575 16574 16574 16574 16574 16574 16574 16574
## [365] 16574 16575 16575 16575 16575 16575 16575 16575 16575 16575 16575 16574 16575
## [378] 16574 16574 16574 16574 16574 16574 16574 16576 16576 16576 16576</code></pre>
<div class="sourceCode" id="cb218"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb218-1"><a href="coalescent.html#cb218-1"></a><span class="co">## note the strings have different length</span></span>
<span id="cb218-2"><a href="coalescent.html#cb218-2"></a><span class="co">## because there is some header information</span></span>
<span id="cb218-3"><a href="coalescent.html#cb218-3"></a><span class="co">## and then because the labels are mixed with the sequences</span></span>
<span id="cb218-4"><a href="coalescent.html#cb218-4"></a></span>
<span id="cb218-5"><a href="coalescent.html#cb218-5"></a><span class="co">## separate out the sequences</span></span>
<span id="cb218-6"><a href="coalescent.html#cb218-6"></a>xx =<span class="st"> </span>x[<span class="kw">nchar</span>(x) <span class="op">&gt;</span><span class="st"> </span><span class="dv">10000</span>] <span class="co">## elements that contain both label and sequences</span></span>
<span id="cb218-7"><a href="coalescent.html#cb218-7"></a>xx.list =<span class="st"> </span><span class="kw">strsplit</span>(xx, <span class="dt">split =</span> <span class="st">&quot;,&quot;</span>) <span class="co">## list of elements, split into label and sequences</span></span>
<span id="cb218-8"><a href="coalescent.html#cb218-8"></a></span>
<span id="cb218-9"><a href="coalescent.html#cb218-9"></a><span class="co">## now split up this list into a vector of labels and a vector of sequences</span></span>
<span id="cb218-10"><a href="coalescent.html#cb218-10"></a>get.first.element =<span class="st"> </span><span class="cf">function</span>(x) {x[<span class="dv">1</span>]}</span>
<span id="cb218-11"><a href="coalescent.html#cb218-11"></a>get.second.element =<span class="st"> </span><span class="cf">function</span>(x) {x[<span class="dv">2</span>]}</span>
<span id="cb218-12"><a href="coalescent.html#cb218-12"></a>labels =<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>(xx.list, get.first.element))</span>
<span id="cb218-13"><a href="coalescent.html#cb218-13"></a>seqs =<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>(xx.list, get.second.element))</span>
<span id="cb218-14"><a href="coalescent.html#cb218-14"></a></span>
<span id="cb218-15"><a href="coalescent.html#cb218-15"></a><span class="co">## Now use labels to select out the 20 Greeks</span></span>
<span id="cb218-16"><a href="coalescent.html#cb218-16"></a>s &lt;-<span class="st"> </span><span class="kw">grepl</span>(<span class="st">&quot;^gre&quot;</span>, labels)</span>
<span id="cb218-17"><a href="coalescent.html#cb218-17"></a>my.labels =<span class="st"> </span>labels[s]</span>
<span id="cb218-18"><a href="coalescent.html#cb218-18"></a></span>
<span id="cb218-19"><a href="coalescent.html#cb218-19"></a>my.seqs =<span class="st"> </span>seqs[s]</span>
<span id="cb218-20"><a href="coalescent.html#cb218-20"></a><span class="kw">nchar</span>(my.seqs) <span class="co">## note the sequences all have same number of bases.</span></span></code></pre></div>
<pre><code>##  [1] 16568 16568 16568 16568 16568 16568 16568 16568 16568 16568 16568 16568 16568
## [14] 16568 16568 16568 16568 16568 16568 16568</code></pre>
<div class="sourceCode" id="cb220"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb220-1"><a href="coalescent.html#cb220-1"></a><span class="co">## now put the bases in a matrix, with each column an indiviual and</span></span>
<span id="cb220-2"><a href="coalescent.html#cb220-2"></a><span class="co">## each row a base.</span></span>
<span id="cb220-3"><a href="coalescent.html#cb220-3"></a>my.list &lt;-<span class="st"> </span><span class="kw">strsplit</span>(my.seqs, <span class="st">&quot;&quot;</span>)</span>
<span id="cb220-4"><a href="coalescent.html#cb220-4"></a>A &lt;-<span class="st"> </span><span class="kw">do.call</span>(cbind, my.list)</span>
<span id="cb220-5"><a href="coalescent.html#cb220-5"></a><span class="kw">dim</span>(A)</span></code></pre></div>
<pre><code>## [1] 16568    20</code></pre>
<div class="sourceCode" id="cb222"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb222-1"><a href="coalescent.html#cb222-1"></a><span class="co">## coding region sequences, 576-16023 (according to &quot;Tree construction</span></span>
<span id="cb222-2"><a href="coalescent.html#cb222-2"></a><span class="co">## and haplogroup prediction&quot; section, but not clear if this was used for</span></span>
<span id="cb222-3"><a href="coalescent.html#cb222-3"></a><span class="co">## Intrapopulation diversity</span></span>
<span id="cb222-4"><a href="coalescent.html#cb222-4"></a>B &lt;-<span class="st"> </span>A[<span class="dv">576</span><span class="op">:</span><span class="dv">16023</span>,]</span>
<span id="cb222-5"><a href="coalescent.html#cb222-5"></a>haps &lt;-<span class="st"> </span>seqs</span></code></pre></div>
<p>Let’s inspect just a bit of one of these sequences</p>
<div class="sourceCode" id="cb223"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb223-1"><a href="coalescent.html#cb223-1"></a><span class="kw">print</span>(<span class="kw">nchar</span>(haps[<span class="dv">213</span>]))</span></code></pre></div>
<pre><code>## [1] 16568</code></pre>
<div class="sourceCode" id="cb225"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb225-1"><a href="coalescent.html#cb225-1"></a>a_segment =<span class="st"> </span><span class="kw">substr</span>(haps[<span class="dv">213</span>], <span class="dv">1</span>, <span class="dv">100</span>)</span>
<span id="cb225-2"><a href="coalescent.html#cb225-2"></a><span class="kw">print</span>(a_segment)</span></code></pre></div>
<pre><code>## [1] &quot;GATCACAGGTCTATCACCCTATTAACCACTCACGGGAGCTCTCCATGCATTTGGTATTTTCGTCTGGGGGGTGTGCACGCGATAGCATTGCGAGACGCTG&quot;</code></pre>
<div class="sourceCode" id="cb227"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb227-1"><a href="coalescent.html#cb227-1"></a><span class="co">## tip: don&#39;t try to print the 16,000 character whole string. it will clog up your computer.</span></span></code></pre></div>
<p>Let’s put the haps in a matrix</p>
<div class="sourceCode" id="cb228"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb228-1"><a href="coalescent.html#cb228-1"></a>my.list &lt;-<span class="st"> </span><span class="kw">strsplit</span>(haps, <span class="st">&quot;&quot;</span>)</span>
<span id="cb228-2"><a href="coalescent.html#cb228-2"></a>H &lt;-<span class="st"> </span><span class="kw">do.call</span>(cbind, my.list)</span>
<span id="cb228-3"><a href="coalescent.html#cb228-3"></a><span class="kw">print</span>(H[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>]) <span class="co">## all the same</span></span></code></pre></div>
<pre><code>##       [,1] [,2] [,3] [,4]
##  [1,] &quot;G&quot;  &quot;G&quot;  &quot;G&quot;  &quot;G&quot; 
##  [2,] &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot; 
##  [3,] &quot;T&quot;  &quot;T&quot;  &quot;T&quot;  &quot;T&quot; 
##  [4,] &quot;C&quot;  &quot;C&quot;  &quot;C&quot;  &quot;C&quot; 
##  [5,] &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot; 
##  [6,] &quot;C&quot;  &quot;C&quot;  &quot;C&quot;  &quot;C&quot; 
##  [7,] &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot; 
##  [8,] &quot;G&quot;  &quot;G&quot;  &quot;G&quot;  &quot;G&quot; 
##  [9,] &quot;G&quot;  &quot;G&quot;  &quot;G&quot;  &quot;G&quot; 
## [10,] &quot;T&quot;  &quot;T&quot;  &quot;T&quot;  &quot;T&quot;</code></pre>
<p>Let’s find a site where there’s polymorphism</p>
<div class="sourceCode" id="cb230"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb230-1"><a href="coalescent.html#cb230-1"></a>hap1 =<span class="st"> </span>H[,<span class="dv">1</span>]</span>
<span id="cb230-2"><a href="coalescent.html#cb230-2"></a>hap2 =<span class="st"> </span>H[,<span class="dv">2</span>]</span>
<span id="cb230-3"><a href="coalescent.html#cb230-3"></a>s &lt;-<span class="st"> </span><span class="kw">min</span>(<span class="kw">which</span>(hap1 <span class="op">!=</span><span class="st"> </span>hap2))</span>
<span id="cb230-4"><a href="coalescent.html#cb230-4"></a><span class="kw">head</span>(H[s <span class="op">+</span><span class="st"> </span><span class="dv">-2</span><span class="op">:</span><span class="dv">2</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>]) <span class="co">## the polymorphic site in context</span></span></code></pre></div>
<pre><code>##      [,1] [,2] [,3] [,4]
## [1,] &quot;T&quot;  &quot;T&quot;  &quot;T&quot;  &quot;T&quot; 
## [2,] &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot; 
## [3,] &quot;A&quot;  &quot;G&quot;  &quot;G&quot;  &quot;G&quot; 
## [4,] &quot;G&quot;  &quot;G&quot;  &quot;G&quot;  &quot;G&quot; 
## [5,] &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;</code></pre>
<p>Let’s see if that’s hap1 is the only “A”</p>
<div class="sourceCode" id="cb232"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb232-1"><a href="coalescent.html#cb232-1"></a><span class="kw">table</span>(H[s,])</span></code></pre></div>
<pre><code>## 
##   A   G 
##  19 361</code></pre>
<p>We see that there are 19 individuals with this “A” instead of “G”.</p>
<blockquote>
<p>Q: How many pairwise differences in total are there between hap1 and
hap2?</p>
</blockquote>
<p>These are the kind of comparisons <em>we</em> will be doing.</p>
</div>
</div>
<div id="ancient-population-estimates" class="section level3">
<h3><span class="header-section-number">10.7.2</span> Ancient population estimates</h3>
Let’s look at Figure 2 on page 5
<div class="figure" style="text-align: center"><span id="fig:batini1"></span>
<img src="figures/batini1.png" alt=" Figure 2 page 5 from @batini2017population" width="80%" />
<p class="caption">
Figure 10.3:  Figure 2 page 5 from <span class="citation">Batini et al. (<a href="#ref-batini2017population" role="doc-biblioref">2017</a>)</span>
</p>
</div>
<ul>
<li>If we look at Ireland (IRE), the think lines are effective population estimates over thousands of years (KYA). For instance, according to mtDNA, 1 thousand years ago the effective population size was around <span class="math inline">\(12,500\)</span>, while 50 thousand years ago the population was around <span class="math inline">\(\approx 1,250\)</span>. The order of magnitude for each these populations appears to be about 10^4 in last few KYA and 10^3 50 KYA. Together, perhaps the size is 10 fold. So we’re looking at European effective population sizes on the order of 100,000 the last few thousand years and on the order of 10,000 tens of thousands of years ago.</li>
</ul>
<div id="using-the-coalescent-to-estimate-changing-population-size" class="section level4">
<h4><span class="header-section-number">10.7.2.1</span> Using the Coalescent to estimate changing population size</h4>
<p>Our procedure will involve a four steps:</p>
<ol style="list-style-type: decimal">
<li>Pick 100 <em>pairs</em> of people at random and count their pairwise differences</li>
<li>Estimate 100 different times of MRCA (<span class="math inline">\(T\)</span>) using assumed mutation rate</li>
<li>Estimate <span class="math inline">\(h(t)\)</span>, the time-varying hazard of coalescence</li>
<li>Estimate the population size $N_e(t)</li>
</ol>
<div id="step-1-random-sampling-and-pairwise-differences" class="section level5">
<h5><span class="header-section-number">10.7.2.1.1</span> Step 1: random sampling and pairwise differences</h5>
<ul>
<li>First, we sample 100 people with replacement.</li>
</ul>
<div class="sourceCode" id="cb234"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb234-1"><a href="coalescent.html#cb234-1"></a><span class="kw">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb234-2"><a href="coalescent.html#cb234-2"></a>hap_ids =<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(H)</span>
<span id="cb234-3"><a href="coalescent.html#cb234-3"></a>hap_id_sample =<span class="st"> </span><span class="kw">sample</span>(hap_ids,</span>
<span id="cb234-4"><a href="coalescent.html#cb234-4"></a>                       <span class="dt">size =</span> <span class="dv">200</span>,</span>
<span id="cb234-5"><a href="coalescent.html#cb234-5"></a>                       <span class="dt">replace =</span> <span class="ot">FALSE</span>)</span>
<span id="cb234-6"><a href="coalescent.html#cb234-6"></a></span>
<span id="cb234-7"><a href="coalescent.html#cb234-7"></a>hap_id.mat &lt;-<span class="st"> </span><span class="kw">matrix</span>(hap_id_sample, <span class="dv">100</span>, <span class="dv">2</span>) <span class="co"># 100 by 2 </span></span>
<span id="cb234-8"><a href="coalescent.html#cb234-8"></a></span>
<span id="cb234-9"><a href="coalescent.html#cb234-9"></a>pairwise_diff_fun &lt;-<span class="st"> </span><span class="cf">function</span>(hap1, hap2)</span>
<span id="cb234-10"><a href="coalescent.html#cb234-10"></a>{</span>
<span id="cb234-11"><a href="coalescent.html#cb234-11"></a>    h1 &lt;-<span class="st"> </span>hap1</span>
<span id="cb234-12"><a href="coalescent.html#cb234-12"></a>    h2 &lt;-<span class="st"> </span>hap2</span>
<span id="cb234-13"><a href="coalescent.html#cb234-13"></a><span class="co">##    h1 &lt;- unlist(strsplit(hap1, &quot;&quot;))</span></span>
<span id="cb234-14"><a href="coalescent.html#cb234-14"></a><span class="co">##    h2 &lt;- unlist(strsplit(hap2, &quot;&quot;))</span></span>
<span id="cb234-15"><a href="coalescent.html#cb234-15"></a>    h1[h1 <span class="op">==</span><span class="st"> &quot;N&quot;</span>] &lt;-<span class="st"> </span><span class="ot">NA</span> <span class="co">## note &quot;N&quot; means missing</span></span>
<span id="cb234-16"><a href="coalescent.html#cb234-16"></a>    h2[h2 <span class="op">==</span><span class="st"> &quot;N&quot;</span>] &lt;-<span class="st"> </span><span class="ot">NA</span> <span class="co">## making these NA avoids counting as polymorphism</span></span>
<span id="cb234-17"><a href="coalescent.html#cb234-17"></a>    k =<span class="st"> </span><span class="kw">sum</span>(h1 <span class="op">!=</span><span class="st"> </span>h2, <span class="dt">na.rm =</span> T)</span>
<span id="cb234-18"><a href="coalescent.html#cb234-18"></a>    n_valid =<span class="st"> </span><span class="kw">sum</span>(<span class="op">!</span><span class="kw">is.na</span>(h1) <span class="op">&amp;</span><span class="st"> </span><span class="op">!</span><span class="kw">is.na</span>(h2))</span>
<span id="cb234-19"><a href="coalescent.html#cb234-19"></a>    <span class="kw">return</span>(<span class="kw">list</span>(<span class="dt">k =</span> k, <span class="dt">n_valid =</span> n_valid))</span>
<span id="cb234-20"><a href="coalescent.html#cb234-20"></a>}</span>
<span id="cb234-21"><a href="coalescent.html#cb234-21"></a></span>
<span id="cb234-22"><a href="coalescent.html#cb234-22"></a><span class="kw">pairwise_diff_fun</span>(H[,<span class="dv">1</span>], H[,<span class="dv">2</span>])</span></code></pre></div>
<pre><code>## $k
## [1] 45
## 
## $n_valid
## [1] 16565</code></pre>
<ul>
<li>Now we’re ready to do pairwise comparisons of all 100 pairs of haplotypes.
<ul>
<li>We’ll define the fraction of locii that have mutated (the pairwise differences) as
<strong>NOTE FOR JOSH: Not sure what P or C are here.</strong>
<span class="math display">\[
\bar{Y} = P/C
\]</span></li>
</ul></li>
</ul>
<div class="sourceCode" id="cb236"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb236-1"><a href="coalescent.html#cb236-1"></a>P.vec =<span class="st"> </span><span class="ot">NULL</span></span>
<span id="cb236-2"><a href="coalescent.html#cb236-2"></a>C.vec =<span class="st"> </span><span class="ot">NULL</span></span>
<span id="cb236-3"><a href="coalescent.html#cb236-3"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(hap_id.mat))</span>
<span id="cb236-4"><a href="coalescent.html#cb236-4"></a>{</span>
<span id="cb236-5"><a href="coalescent.html#cb236-5"></a>    hap_id<span class="fl">.1</span> =<span class="st"> </span>hap_id.mat[i,<span class="dv">1</span>]</span>
<span id="cb236-6"><a href="coalescent.html#cb236-6"></a>    hap_id<span class="fl">.2</span> =<span class="st"> </span>hap_id.mat[i,<span class="dv">2</span>]</span>
<span id="cb236-7"><a href="coalescent.html#cb236-7"></a>    hap1 =<span class="st"> </span>H[,hap_id<span class="fl">.1</span>]</span>
<span id="cb236-8"><a href="coalescent.html#cb236-8"></a>    hap2 =<span class="st"> </span>H[,hap_id<span class="fl">.2</span>]</span>
<span id="cb236-9"><a href="coalescent.html#cb236-9"></a>    out =<span class="st"> </span><span class="kw">pairwise_diff_fun</span>(hap1, hap2)</span>
<span id="cb236-10"><a href="coalescent.html#cb236-10"></a>    P.vec[i] =<span class="st"> </span>out<span class="op">$</span>k</span>
<span id="cb236-11"><a href="coalescent.html#cb236-11"></a>    C.vec[i] =<span class="st"> </span>out<span class="op">$</span>n_valid</span>
<span id="cb236-12"><a href="coalescent.html#cb236-12"></a>}</span>
<span id="cb236-13"><a href="coalescent.html#cb236-13"></a></span>
<span id="cb236-14"><a href="coalescent.html#cb236-14"></a>Y.bar =<span class="st"> </span>P.vec<span class="op">/</span>C.vec</span>
<span id="cb236-15"><a href="coalescent.html#cb236-15"></a><span class="kw">head</span>(P.vec)</span></code></pre></div>
<pre><code>## [1] 19 36 13 39  5 84</code></pre>
<div class="sourceCode" id="cb238"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb238-1"><a href="coalescent.html#cb238-1"></a><span class="kw">head</span>(C.vec)</span></code></pre></div>
<pre><code>## [1] 16566 16484 16539 16565 16565 16544</code></pre>
<div class="sourceCode" id="cb240"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb240-1"><a href="coalescent.html#cb240-1"></a><span class="kw">head</span>(Y.bar)</span></code></pre></div>
<pre><code>## [1] 0.0011469274 0.0021839359 0.0007860209 0.0023543616 0.0003018412 0.0050773694</code></pre>
</div>
<div id="step-2-use-mutation-rate-to-obtain-100-distinct-tmrca" class="section level5">
<h5><span class="header-section-number">10.7.2.1.2</span> Step 2: Use mutation rate to obtain 100 distinct TMRCA</h5>
<ul>
<li>Now we’ll use the mutation rate <span class="math inline">\(\theta_m\)</span> given by <span class="citation">Batini et al. (<a href="#ref-batini2017population" role="doc-biblioref">2017</a>)</span> to compute time back to MRCA.
<ul>
<li>First, some background:
<ul>
<li>Let <span class="math inline">\(a\)</span> index sites and write <span class="math inline">\(Y_a = 1\)</span> if the letters are different, and <span class="math inline">\(Y_a = 0\)</span> otherwise.</li>
<li>For 1 site, the probability that it there has been a mutation, is 1 minus the chance that there has been no mutation.</li>
</ul></li>
</ul></li>
</ul>
<p><span class="math display">\[
     P( Y_a = 1 ) = 1 - e^{-T\theta}
\]</span></p>
<ul>
<li>The mean <span class="math inline">\(\bar{Y}\)</span> across the segment is the proportion of <span class="math inline">\(Y_a\)</span> that equal 1. So, in expectation for the fraction of sites that mutate is the same as the probability that 1 site mutates (assuming independence of mutation probabilities by site). This allows us to write:</li>
</ul>
<p><span class="math display">\[
     E{\bar{Y}} = 1 - e^{-2T\theta},
\]</span></p>
<ul>
<li>where we’ve added a “2” in order to account that either one of the pairwise branched could have had a mutation, so we our “exposure” is twice the time to MRCA.
<ul>
<li><p>Rearranging we get an estimate <span class="math inline">\(\hat{T}\)</span> of <span class="math inline">\(T\)</span> to be
<span class="math display">\[
\hat{T} = {-\log (1 - \bar{Y} ) \over \theta}
\]</span></p></li>
<li><p>Now we’re ready to estimate the MRCAs</p></li>
</ul></li>
</ul>
<div class="sourceCode" id="cb242"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb242-1"><a href="coalescent.html#cb242-1"></a>theta_m =<span class="st"> </span><span class="fl">2.21</span> <span class="op">*</span><span class="st"> </span><span class="dv">10</span><span class="op">^</span>(<span class="op">-</span><span class="dv">8</span>) <span class="co">## Batini page 6 (TMRCA estimation)</span></span>
<span id="cb242-2"><a href="coalescent.html#cb242-2"></a>T.vec &lt;-<span class="st"> </span><span class="op">-</span>(<span class="dv">1</span><span class="op">/</span><span class="dv">2</span>) <span class="op">*</span><span class="st"> </span>(<span class="dv">1</span><span class="op">/</span>theta_m) <span class="op">*</span><span class="st"> </span><span class="kw">log</span>(<span class="dv">1</span> <span class="op">-</span><span class="st"> </span>Y.bar) <span class="co">## TMRCAs in years ago</span></span>
<span id="cb242-3"><a href="coalescent.html#cb242-3"></a><span class="kw">head</span>(T.vec, <span class="dt">n =</span> <span class="dv">10</span>)</span></code></pre></div>
<pre><code>##  [1]  25963.477  49464.349  17790.271  53328.902   6830.018 115165.228  56610.650
##  [8]  12294.775  21863.319  54701.270</code></pre>
<ul>
<li>Let’s visualize these</li>
</ul>
<div class="sourceCode" id="cb244"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb244-1"><a href="coalescent.html#cb244-1"></a><span class="kw">hist</span>(T.vec)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-176-1.png" width="672" /></p>
<ul>
<li>We see a lot of coalescence about 50 KYA, which is as far back as <span class="citation">Batini et al. (<a href="#ref-batini2017population" role="doc-biblioref">2017</a>)</span> estimates go. This means the population was small back then.</li>
</ul>
</div>
<div id="step-3-obtaining-time-varying-hazard-of-coalescence" class="section level5">
<h5><span class="header-section-number">10.7.2.1.3</span> Step 3: Obtaining time-varying hazard of coalescence</h5>
<ul>
<li>We’ll do this in two ways:
<ol style="list-style-type: decimal">
<li>Compute the slope of the logarithm of the survival curve, but we’ll see that it is noisy and needs to be smoothed.</li>
<li>Construct a “life table” of coalescence with discrete periods of time.</li>
</ol>
<ul>
<li>Let’s start with the more continuous version of estimating slopes.</li>
</ul></li>
</ul>
<div class="sourceCode" id="cb245"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb245-1"><a href="coalescent.html#cb245-1"></a><span class="co">## Plot survival curve by order of T</span></span>
<span id="cb245-2"><a href="coalescent.html#cb245-2"></a>St =<span class="st"> </span>(<span class="dv">100</span><span class="op">:</span><span class="dv">1</span>)<span class="op">/</span><span class="dv">100</span> <span class="co">## or more generally (length(T.vec):1)/length(T.vec)</span></span>
<span id="cb245-3"><a href="coalescent.html#cb245-3"></a>t =<span class="st"> </span>kya =<span class="st"> </span><span class="kw">sort</span>(T.vec)<span class="op">/</span><span class="dv">1000</span></span>
<span id="cb245-4"><a href="coalescent.html#cb245-4"></a><span class="kw">plot</span>(kya, St, <span class="dt">type =</span> <span class="st">&quot;l&quot;</span>,</span>
<span id="cb245-5"><a href="coalescent.html#cb245-5"></a>     <span class="dt">xlab =</span> <span class="st">&quot;Kilo Years Ago&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Fraction of pairs without common ancestor&quot;</span>,</span>
<span id="cb245-6"><a href="coalescent.html#cb245-6"></a>     <span class="dt">main =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-177-1"></span>
<img src="bookdown-demo_files/figure-html/unnamed-chunk-177-1.png" alt="Estimatated probability of not coalescing" width="672" />
<p class="caption">
Figure 10.4: Estimatated probability of not coalescing
</p>
</div>
<div class="sourceCode" id="cb246"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb246-1"><a href="coalescent.html#cb246-1"></a><span class="kw">plot</span>(kya, <span class="kw">log</span>(St), <span class="dt">type =</span> <span class="st">&quot;l&quot;</span>,</span>
<span id="cb246-2"><a href="coalescent.html#cb246-2"></a>     <span class="dt">xlab =</span> <span class="st">&quot;Kilo Years Ago&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Log fraction of pairs without common ancestor&quot;</span>,</span>
<span id="cb246-3"><a href="coalescent.html#cb246-3"></a>     <span class="dt">main =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-177-2"></span>
<img src="bookdown-demo_files/figure-html/unnamed-chunk-177-2.png" alt="Estimatated probability of not coalescing" width="672" />
<p class="caption">
Figure 10.5: Estimatated probability of not coalescing
</p>
</div>
<ul>
<li>The slope in the first 50,000 years has increased, which implies that the hazard has also increased. This means that the population has increased over time.</li>
</ul>
<div id="estimating-hazards-method-1.-from-smoothed-survival-curve" class="section level6">
<h6><span class="header-section-number">10.7.2.1.3.1</span> <strong>Estimating Hazards method 1. From smoothed survival curve</strong></h6>
<ul>
<li>First, we smooth the survival curve (St)</li>
</ul>
<div class="sourceCode" id="cb247"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb247-1"><a href="coalescent.html#cb247-1"></a>out =<span class="st"> </span><span class="kw">lowess</span>(<span class="dt">x =</span> kya, <span class="dt">y =</span> St, <span class="dt">f =</span> <span class="dv">1</span><span class="op">/</span><span class="dv">5</span>)</span>
<span id="cb247-2"><a href="coalescent.html#cb247-2"></a></span>
<span id="cb247-3"><a href="coalescent.html#cb247-3"></a>St_smooth =<span class="st"> </span>out<span class="op">$</span>y</span>
<span id="cb247-4"><a href="coalescent.html#cb247-4"></a>kya_smooth =<span class="st"> </span>out<span class="op">$</span>x</span>
<span id="cb247-5"><a href="coalescent.html#cb247-5"></a><span class="kw">plot</span>(kya, St, <span class="dt">cex =</span> <span class="fl">.5</span>, <span class="dt">ylab=</span><span class="st">&quot;Fraction of pairs without common ancestor&quot;</span>,<span class="dt">xlab=</span><span class="st">&quot;Kilo Years Ago&quot;</span>)</span>
<span id="cb247-6"><a href="coalescent.html#cb247-6"></a><span class="kw">lines</span>(kya_smooth, St_smooth, <span class="dt">type =</span> <span class="st">&#39;l&#39;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-178"></span>
<img src="bookdown-demo_files/figure-html/unnamed-chunk-178-1.png" alt="Smoothing" width="672" />
<p class="caption">
Figure 10.6: Smoothing
</p>
</div>
<ul>
<li>In this case we tried an f parameter of 1/5 but it can be modified to get a better fit.
<ul>
<li>To convert the smoothed survivorship curve into hazards, we use the fact that hazards are equal to minus the slope of log survivorship.</li>
</ul></li>
</ul>
<div class="sourceCode" id="cb248"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb248-1"><a href="coalescent.html#cb248-1"></a>haz_hat =<span class="st"> </span><span class="op">-</span><span class="kw">diff</span>(St_smooth)<span class="op">/</span><span class="kw">diff</span>(kya_smooth)</span>
<span id="cb248-2"><a href="coalescent.html#cb248-2"></a><span class="kw">plot</span>(kya_smooth[<span class="op">-</span><span class="dv">1</span>], haz_hat, <span class="dt">type =</span> <span class="st">&#39;l&#39;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Estimated Hazard&quot;</span>, </span>
<span id="cb248-3"><a href="coalescent.html#cb248-3"></a>     <span class="dt">xlab =</span> <span class="st">&quot;Kilo Years Ago (Smooth)&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-179"></span>
<img src="bookdown-demo_files/figure-html/unnamed-chunk-179-1.png" alt="Hazards" width="672" />
<p class="caption">
Figure 10.7: Hazards
</p>
</div>
<ul>
<li>Unfortunately, this plot does not tell us anything about the uncertainty of our estimates.</li>
</ul>
</div>
<div id="estimating-hazards-method-2.-life-table-appraoch" class="section level6">
<h6><span class="header-section-number">10.7.2.1.3.2</span> <strong>Estimating Hazards method 2. Life table appraoch</strong></h6>
<ul>
<li>Now let’s estimate the hazards using a “life table”, where again “death” is coalescence and “survival” is still not having a common ancestor.</li>
</ul>
<div class="sourceCode" id="cb249"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb249-1"><a href="coalescent.html#cb249-1"></a><span class="co">## we choose these time boundaries arbitrarily </span></span>
<span id="cb249-2"><a href="coalescent.html#cb249-2"></a>x =<span class="st"> </span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">2</span>, <span class="dv">5</span>, <span class="dv">10</span>,  <span class="dv">20</span>, <span class="dv">30</span>, <span class="dv">40</span>, <span class="dv">65</span>, <span class="dv">180</span>) <span class="op">*</span><span class="st"> </span><span class="dv">1000</span> <span class="co">## time interval boundaries</span></span></code></pre></div>
<ul>
<li>Define a function to count “exposure” by those pairs that have MRCA in time intervals.</li>
</ul>
<div class="sourceCode" id="cb250"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb250-1"><a href="coalescent.html#cb250-1"></a>get_nax &lt;-<span class="st"> </span><span class="cf">function</span>(Ti, x)</span>
<span id="cb250-2"><a href="coalescent.html#cb250-2"></a>{</span>
<span id="cb250-3"><a href="coalescent.html#cb250-3"></a>    <span class="co">## get person years lived in interval by those who die</span></span>
<span id="cb250-4"><a href="coalescent.html#cb250-4"></a>    nax &lt;-<span class="st"> </span><span class="ot">NULL</span></span>
<span id="cb250-5"><a href="coalescent.html#cb250-5"></a>    <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>(<span class="kw">length</span>(x)<span class="op">-</span><span class="dv">1</span>))</span>
<span id="cb250-6"><a href="coalescent.html#cb250-6"></a>    {</span>
<span id="cb250-7"><a href="coalescent.html#cb250-7"></a>        s &lt;-<span class="st"> </span>Ti <span class="op">&gt;=</span><span class="st"> </span>x[i] <span class="op">&amp;</span><span class="st"> </span>Ti <span class="op">&lt;</span><span class="st"> </span>x[i<span class="op">+</span><span class="dv">1</span>]</span>
<span id="cb250-8"><a href="coalescent.html#cb250-8"></a>        <span class="cf">if</span> (<span class="kw">length</span>(Ti[s]) <span class="op">!=</span><span class="st"> </span><span class="dv">0</span>) {</span>
<span id="cb250-9"><a href="coalescent.html#cb250-9"></a>          nax[i] =<span class="st"> </span><span class="kw">mean</span>(Ti[s] <span class="op">-</span><span class="st"> </span>x[i])</span>
<span id="cb250-10"><a href="coalescent.html#cb250-10"></a>        }</span>
<span id="cb250-11"><a href="coalescent.html#cb250-11"></a>        nax[<span class="kw">is.na</span>(nax)] &lt;-<span class="st"> </span><span class="dv">0</span></span>
<span id="cb250-12"><a href="coalescent.html#cb250-12"></a>    }</span>
<span id="cb250-13"><a href="coalescent.html#cb250-13"></a>    <span class="kw">return</span>(nax)</span>
<span id="cb250-14"><a href="coalescent.html#cb250-14"></a>}</span></code></pre></div>
<ul>
<li>Construct the life table</li>
</ul>
<div class="sourceCode" id="cb251"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb251-1"><a href="coalescent.html#cb251-1"></a>n &lt;-<span class="st"> </span><span class="kw">diff</span>(x)</span>
<span id="cb251-2"><a href="coalescent.html#cb251-2"></a>T.vec.by.cat &lt;-<span class="st"> </span><span class="kw">cut</span>(T.vec, x, <span class="dt">include.lowest =</span> T, <span class="dt">right =</span> F)</span>
<span id="cb251-3"><a href="coalescent.html#cb251-3"></a>ndx =<span class="st"> </span><span class="kw">table</span>(T.vec.by.cat)</span>
<span id="cb251-4"><a href="coalescent.html#cb251-4"></a>lx =<span class="st"> </span><span class="kw">rev</span>(<span class="kw">cumsum</span>(<span class="kw">rev</span>(ndx)))</span>
<span id="cb251-5"><a href="coalescent.html#cb251-5"></a>lxpn =<span class="st"> </span><span class="kw">c</span>(lx[<span class="op">-</span><span class="dv">1</span>], <span class="dv">0</span>)</span>
<span id="cb251-6"><a href="coalescent.html#cb251-6"></a>nax =<span class="st"> </span><span class="kw">get_nax</span>(<span class="dt">Ti =</span> T.vec, <span class="dt">x =</span> x)</span>
<span id="cb251-7"><a href="coalescent.html#cb251-7"></a>nLx =<span class="st"> </span>n<span class="op">*</span>lxpn <span class="op">+</span><span class="st"> </span>nax <span class="op">*</span><span class="st"> </span>ndx <span class="co">## exposure</span></span>
<span id="cb251-8"><a href="coalescent.html#cb251-8"></a>nmx =<span class="st"> </span>ndx<span class="op">/</span>nLx <span class="co">## hazard</span></span>
<span id="cb251-9"><a href="coalescent.html#cb251-9"></a>lt &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="dt">x =</span> x[<span class="op">-</span><span class="kw">length</span>(x)], n, ndx, lx, nax, nLx, nmx)</span>
<span id="cb251-10"><a href="coalescent.html#cb251-10"></a><span class="kw">print</span>(lt)</span></code></pre></div>
<pre><code>##                       x      n ndx  lx       nax      nLx          nmx
## [0,2e+03)             0   2000   0 100     0.000 200000.0 0.000000e+00
## [2e+03,5e+03)      2000   3000   0 100     0.000 300000.0 0.000000e+00
## [5e+03,1e+04)      5000   5000   3 100  3651.934 495955.8 6.048926e-06
## [1e+04,2e+04)     10000  10000   7  97  6416.957 944918.7 7.408045e-06
## [2e+04,3e+04)     20000  10000   7  90  3662.324 855636.3 8.181046e-06
## [3e+04,4e+04)     30000  10000  19  83  6798.251 769166.8 2.470206e-05
## [4e+04,6.5e+04)   40000  25000  54  64 11619.224 877438.1 6.154280e-05
## [6.5e+04,1.8e+05] 65000 115000  10  10 24611.455 246114.6 4.063149e-05</code></pre>
<ul>
<li>Let’s compare the two estimates:</li>
</ul>
<div class="sourceCode" id="cb253"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb253-1"><a href="coalescent.html#cb253-1"></a>x.mid =<span class="st"> </span>x[<span class="op">-</span><span class="kw">length</span>(x)] <span class="op">+</span><span class="st"> </span>n<span class="op">/</span><span class="dv">2</span></span>
<span id="cb253-2"><a href="coalescent.html#cb253-2"></a><span class="kw">plot</span>(x.mid, nmx, <span class="dt">type =</span> <span class="st">&#39;o&#39;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Mortality Rate&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;Age (Midpoint)&quot;</span>, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>)</span>
<span id="cb253-3"><a href="coalescent.html#cb253-3"></a><span class="kw">axis</span>(<span class="dv">2</span>)</span>
<span id="cb253-4"><a href="coalescent.html#cb253-4"></a><span class="kw">lines</span>(kya_smooth[<span class="op">-</span><span class="dv">1</span>] <span class="op">*</span><span class="st"> </span><span class="dv">1000</span>, haz_hat<span class="op">/</span><span class="dv">1000</span>, <span class="dt">type =</span> <span class="st">&quot;l&quot;</span>, <span class="dt">col=</span><span class="st">&quot;blue&quot;</span>)</span>
<span id="cb253-5"><a href="coalescent.html#cb253-5"></a><span class="kw">legend</span>(<span class="st">&quot;topleft&quot;</span> , <span class="dt">legend=</span><span class="kw">c</span>(<span class="st">&quot;Lifetable&quot;</span>, <span class="st">&quot;Smooth Survival Curve&quot;</span>),</span>
<span id="cb253-6"><a href="coalescent.html#cb253-6"></a>       <span class="dt">col=</span><span class="kw">c</span>(<span class="st">&quot;red&quot;</span>, <span class="st">&quot;blue&quot;</span>), <span class="dt">lty=</span><span class="dv">1</span><span class="op">:</span><span class="dv">2</span>, <span class="dt">cex=</span><span class="fl">0.8</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-183"></span>
<img src="bookdown-demo_files/figure-html/unnamed-chunk-183-1.png" alt="Hazards comparison" width="672" />
<p class="caption">
Figure 10.8: Hazards comparison
</p>
</div>
<ul>
<li>We’re getting basically the same thing, with a little more hint of rising hazards (shrinking pop size) in first 20 kya. Nearly the same thing estimate 40 kya ago, and very little signal before that.</li>
</ul>
</div>
</div>
<div id="step-4-estimate-the-population-size-n_et" class="section level5">
<h5><span class="header-section-number">10.7.2.1.4</span> Step 4: Estimate the population size <span class="math inline">\(N_e(t)\)</span></h5>
<ul>
<li>In order to estimate the population size, we have to think for a minute about units.
<ul>
<li>Our hazards are per year, but our logic for why hazards are related to population size is per generation. Remember, the chance of coalescence <em>per generation</em> was <span class="math inline">\(1/N\)</span>. This means we will want to multiply the annual hazard by generation length in order to get hazards per generation unit of time.</li>
<li>Second, mtDNA is inherited only through mothers. So the coalescent that we are thinking of is only for women. And the effective population size we’re estimating is for females only. We can get a rough estimate of both sexes by doubling the number of females.</li>
<li>Putting these two considerations together, we have:
<span class="math display">\[
\hat{N_e(both sexes)} = 2 \times {1 \over 25 \cdot h(t)},
\]</span></li>
<li>The 2 inflates to both sexes, and the 25 inflates the annual hazard into generations of 25 years in length.</li>
</ul></li>
</ul>
<div class="sourceCode" id="cb254"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb254-1"><a href="coalescent.html#cb254-1"></a>Ne_smooth =<span class="st"> </span><span class="dv">2</span> <span class="op">/</span><span class="st"> </span>(haz_hat<span class="op">/</span><span class="dv">1000</span> <span class="op">*</span><span class="st"> </span><span class="dv">25</span>)</span>
<span id="cb254-2"><a href="coalescent.html#cb254-2"></a></span>
<span id="cb254-3"><a href="coalescent.html#cb254-3"></a>Ne_lifetable =<span class="st"> </span><span class="dv">2</span> <span class="op">/</span><span class="st"> </span>(nmx <span class="op">*</span><span class="st"> </span><span class="dv">25</span>)</span>
<span id="cb254-4"><a href="coalescent.html#cb254-4"></a></span>
<span id="cb254-5"><a href="coalescent.html#cb254-5"></a><span class="co">## create step function for plotting</span></span>
<span id="cb254-6"><a href="coalescent.html#cb254-6"></a>Ne_lifetable_step =<span class="st"> </span><span class="kw">rep</span>(Ne_lifetable, n<span class="op">/</span><span class="dv">1000</span>)</span>
<span id="cb254-7"><a href="coalescent.html#cb254-7"></a>kya_step =<span class="st"> </span><span class="dv">1</span><span class="op">:</span>(<span class="kw">max</span>(x)<span class="op">/</span><span class="dv">1000</span>)</span></code></pre></div>
<ul>
<li>Plotting the results</li>
</ul>
<div class="sourceCode" id="cb255"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb255-1"><a href="coalescent.html#cb255-1"></a><span class="kw">plot</span>(kya_smooth[<span class="op">-</span><span class="dv">1</span>], Ne_smooth, <span class="dt">type =</span> <span class="st">&#39;l&#39;</span>, <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="dv">1000</span>, <span class="dv">60000</span>), <span class="dt">log =</span> <span class="st">&#39;y&#39;</span>,</span>
<span id="cb255-2"><a href="coalescent.html#cb255-2"></a>     <span class="dt">lty =</span> <span class="dv">2</span>, <span class="dt">ylab=</span><span class="st">&quot;Estimated Population Size&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;Kilo Years Ago (Smooth)&quot;</span>)</span>
<span id="cb255-3"><a href="coalescent.html#cb255-3"></a><span class="kw">lines</span>(kya_step, Ne_lifetable_step, <span class="dt">type =</span> <span class="st">&#39;l&#39;</span>, <span class="dt">lwd =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-185-1.png" width="672" /></p>
<ul>
<li>We could evaluate uncertainty using resampling techniques such as bootstrap.
<ul>
<li>The estimate above doesn’t include the fact that maybe the mutation rate is not constant or that the sample of people is not representative.
<ul>
<li>Also, it’s possible that we are not seeing mutations that are not sampled.</li>
<li>Migration: maybe the people are not all the same population.</li>
</ul></li>
<li>Let’s loop through and do the whole estimation 40 times</li>
</ul></li>
</ul>
<div class="sourceCode" id="cb256"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb256-1"><a href="coalescent.html#cb256-1"></a>n_trials =<span class="st"> </span><span class="dv">40</span></span>
<span id="cb256-2"><a href="coalescent.html#cb256-2"></a>Ne.mat &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="ot">NA</span>, <span class="dt">nrow =</span> n_trials, <span class="dt">ncol =</span> <span class="kw">max</span>(x)<span class="op">/</span><span class="dv">1000</span>)</span>
<span id="cb256-3"><a href="coalescent.html#cb256-3"></a><span class="kw">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb256-4"><a href="coalescent.html#cb256-4"></a><span class="cf">for</span> (r <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>n_trials)</span>
<span id="cb256-5"><a href="coalescent.html#cb256-5"></a>{</span>
<span id="cb256-6"><a href="coalescent.html#cb256-6"></a>  <span class="co">#r = 1</span></span>
<span id="cb256-7"><a href="coalescent.html#cb256-7"></a>    <span class="co">#print(r)</span></span>
<span id="cb256-8"><a href="coalescent.html#cb256-8"></a>    <span class="co">## sample</span></span>
<span id="cb256-9"><a href="coalescent.html#cb256-9"></a>    hap_id_sample =<span class="st"> </span><span class="kw">sample</span>(hap_ids,</span>
<span id="cb256-10"><a href="coalescent.html#cb256-10"></a>                           <span class="dt">size =</span> <span class="dv">200</span>,</span>
<span id="cb256-11"><a href="coalescent.html#cb256-11"></a>                           <span class="dt">replace =</span> <span class="ot">FALSE</span>)</span>
<span id="cb256-12"><a href="coalescent.html#cb256-12"></a>    hap_id.mat &lt;-<span class="st"> </span><span class="kw">matrix</span>(hap_id_sample, <span class="dv">100</span>, <span class="dv">2</span>)</span>
<span id="cb256-13"><a href="coalescent.html#cb256-13"></a>    <span class="co">## estimate MRCA distribution</span></span>
<span id="cb256-14"><a href="coalescent.html#cb256-14"></a>    P.vec =<span class="st"> </span><span class="ot">NULL</span></span>
<span id="cb256-15"><a href="coalescent.html#cb256-15"></a>    C.vec =<span class="st"> </span><span class="ot">NULL</span></span>
<span id="cb256-16"><a href="coalescent.html#cb256-16"></a>    <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(hap_id.mat))</span>
<span id="cb256-17"><a href="coalescent.html#cb256-17"></a>    {</span>
<span id="cb256-18"><a href="coalescent.html#cb256-18"></a>        hap_id<span class="fl">.1</span> =<span class="st"> </span>hap_id.mat[i,<span class="dv">1</span>]</span>
<span id="cb256-19"><a href="coalescent.html#cb256-19"></a>        hap_id<span class="fl">.2</span> =<span class="st"> </span>hap_id.mat[i,<span class="dv">2</span>]</span>
<span id="cb256-20"><a href="coalescent.html#cb256-20"></a>        hap1 =<span class="st"> </span>H[,hap_id<span class="fl">.1</span>]</span>
<span id="cb256-21"><a href="coalescent.html#cb256-21"></a>        hap2 =<span class="st"> </span>H[,hap_id<span class="fl">.2</span>]</span>
<span id="cb256-22"><a href="coalescent.html#cb256-22"></a>        out =<span class="st"> </span><span class="kw">pairwise_diff_fun</span>(hap1, hap2)</span>
<span id="cb256-23"><a href="coalescent.html#cb256-23"></a>        P.vec[i] =<span class="st"> </span>out<span class="op">$</span>k</span>
<span id="cb256-24"><a href="coalescent.html#cb256-24"></a>        C.vec[i] =<span class="st"> </span>out<span class="op">$</span>n_valid</span>
<span id="cb256-25"><a href="coalescent.html#cb256-25"></a>    }</span>
<span id="cb256-26"><a href="coalescent.html#cb256-26"></a>    Y.bar =<span class="st"> </span>P.vec<span class="op">/</span>C.vec</span>
<span id="cb256-27"><a href="coalescent.html#cb256-27"></a>    T.vec &lt;-<span class="st"> </span><span class="op">-</span>(<span class="dv">1</span><span class="op">/</span><span class="dv">2</span>) <span class="op">*</span><span class="st"> </span>(<span class="dv">1</span><span class="op">/</span>theta_m) <span class="op">*</span><span class="st"> </span><span class="kw">log</span>(<span class="dv">1</span> <span class="op">-</span><span class="st"> </span>Y.bar) <span class="co">## TMRCAs in years ago</span></span>
<span id="cb256-28"><a href="coalescent.html#cb256-28"></a>    <span class="co">## estimate Ne</span></span>
<span id="cb256-29"><a href="coalescent.html#cb256-29"></a>    T.vec.by.cat &lt;-<span class="st"> </span><span class="kw">cut</span>(T.vec, x, <span class="dt">include.lowest =</span> T, <span class="dt">right =</span> F)</span>
<span id="cb256-30"><a href="coalescent.html#cb256-30"></a>    ndx =<span class="st"> </span><span class="kw">table</span>(T.vec.by.cat)</span>
<span id="cb256-31"><a href="coalescent.html#cb256-31"></a>    lx =<span class="st"> </span><span class="kw">rev</span>(<span class="kw">cumsum</span>(<span class="kw">rev</span>(ndx)))</span>
<span id="cb256-32"><a href="coalescent.html#cb256-32"></a>    lxpn =<span class="st"> </span><span class="kw">c</span>(lx[<span class="op">-</span><span class="dv">1</span>], <span class="dv">0</span>)</span>
<span id="cb256-33"><a href="coalescent.html#cb256-33"></a>    nax =<span class="st"> </span><span class="kw">get_nax</span>(<span class="dt">Ti =</span> T.vec, <span class="dt">x =</span> x)</span>
<span id="cb256-34"><a href="coalescent.html#cb256-34"></a>    nLx =<span class="st"> </span>n<span class="op">*</span>lxpn <span class="op">+</span><span class="st"> </span>nax <span class="op">*</span><span class="st"> </span>ndx <span class="co">## exposure</span></span>
<span id="cb256-35"><a href="coalescent.html#cb256-35"></a>    nmx =<span class="st"> </span>ndx<span class="op">/</span>nLx <span class="co">## hazard</span></span>
<span id="cb256-36"><a href="coalescent.html#cb256-36"></a>    Ne_lifetable =<span class="st"> </span><span class="dv">2</span> <span class="op">/</span><span class="st"> </span>(nmx <span class="op">*</span><span class="st"> </span><span class="dv">25</span>)</span>
<span id="cb256-37"><a href="coalescent.html#cb256-37"></a>    <span class="co">## create step function for plotting</span></span>
<span id="cb256-38"><a href="coalescent.html#cb256-38"></a>    Ne_lifetable_step =<span class="st"> </span><span class="kw">rep</span>(Ne_lifetable, n<span class="op">/</span><span class="dv">1000</span>)</span>
<span id="cb256-39"><a href="coalescent.html#cb256-39"></a>    kya_step =<span class="st"> </span><span class="dv">1</span><span class="op">:</span>(<span class="kw">max</span>(x)<span class="op">/</span><span class="dv">1000</span>)</span>
<span id="cb256-40"><a href="coalescent.html#cb256-40"></a>    <span class="co">## save result</span></span>
<span id="cb256-41"><a href="coalescent.html#cb256-41"></a>    Ne.mat[r,] &lt;-<span class="st"> </span>Ne_lifetable_step</span>
<span id="cb256-42"><a href="coalescent.html#cb256-42"></a>}</span>
<span id="cb256-43"><a href="coalescent.html#cb256-43"></a></span>
<span id="cb256-44"><a href="coalescent.html#cb256-44"></a>Ne.interval &lt;-<span class="st"> </span><span class="kw">apply</span>(Ne.mat, <span class="dv">2</span>, quantile, <span class="kw">c</span>(.<span class="dv">1</span>,.<span class="dv">5</span>, <span class="fl">.9</span>))</span>
<span id="cb256-45"><a href="coalescent.html#cb256-45"></a><span class="kw">matplot</span>(<span class="kw">t</span>(Ne.interval), <span class="dt">type =</span> <span class="st">&#39;l&#39;</span>, <span class="dt">log =</span> <span class="st">&#39;y&#39;</span>, <span class="dt">col =</span> <span class="st">&quot;grey&quot;</span>, <span class="dt">lty =</span> <span class="dv">1</span>, <span class="dt">lwd =</span> <span class="dv">2</span>, <span class="dt">ylab=</span><span class="st">&quot;Estimated Population Size&quot;</span>, <span class="dt">xlab=</span><span class="st">&quot;Kilo Years Ago&quot;</span>)</span>
<span id="cb256-46"><a href="coalescent.html#cb256-46"></a><span class="kw">lines</span>(Ne.interval[<span class="st">&quot;50%&quot;</span>,], <span class="dt">lwd =</span> <span class="dv">4</span>)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-186-1.png" width="672" /></p>
<ul>
<li>So it seems fairly clear that effective population size has been growing the last 50 thousand years, from a low of a few thousand to a few tens of thousand.
<ul>
<li>The general trend in growth is consistent with Batini but
<ul>
<li>Our total population size seems smaller by a factor of about 4 or 5.</li>
<li>We don’t have the resolution to see increase between 10 and 20 kya, the end of the Last Glacial Maximum or the more recent Bronze Age steppe expansion 2 to 5 kya.</li>
</ul></li>
<li>In order to get more resolution and study sub-group differences, we would want to turn to methods that do more than pair-wise comparisons, giving us more detailed information about the effective population sizes of the past.</li>
</ul></li>
</ul>

</div>
</div>
</div>
</div>
</div>
<h3>References</h3>
<div id="refs" class="references">
<div id="ref-batini2017population">
<p>Batini, Chiara, Pille Hallast, Åshild J Vågene, Daniel Zadik, Heidi A Eriksen, Horolma Pamjav, Antti Sajantila, Jon H Wetton, and Mark A Jobling. 2017. “Population Resequencing of European Mitochondrial Genomes Highlights Sex-Bias in Bronze Age Demographic Expansions.” <em>Scientific Reports</em> 7 (1): 1–8.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="fisher-wright.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="student-presentations.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/09-coalescent.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["bookdown-demo.pdf", "bookdown-demo.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
